
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Tutorial 1: Single cell simulation with current injection (with BioNet) &#8212; Brain Modeling Toolkit 1.0.6 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/aibs_sphinx.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "document", "processHtmlClass": "math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 2: Single cell simulation with external feedfoward input (with BioNet)" href="tutorial_single_cell_syn.html" />
    <link rel="prev" title="BMTK Builder (A Quick Introduction)" href="tutorial_NetworkBuilder_Intro.html" /> 
  </head><body>
<script type="text/javascript" src="_static/external_assets/javascript/portal.js"></script>
<style>

	/* the icon font-family definition and helper classes */
	@font-face {
	  font-family: 'portal_icon_font';
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677');
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677#iefix') format('embedded-opentype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.woff?39505677') format('woff'),
	       url('_static/external_assets/stylesheets/portal_icon_font.ttf?39505677') format('truetype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.svg?39505677#portal_icon_font') format('svg');
	  font-weight: normal;
	  font-style: normal;
	}
	 
	 [class^="icon-"]:before, [class*=" icon-"]:before {
	  font-family: "portal_icon_font";
	  font-style: normal;
	  font-weight: normal;
	  speak: none;
	 
	  display: inline-block;
	  text-decoration: inherit;
	  width: 1em;
	  margin-right: .2em;
	  text-align: center;
	  /* opacity: .8; */
	 
	  /* For safety - reset parent styles, that can break glyph codes*/
	  font-variant: normal;
	  text-transform: none;
	     
	  /* fix buttons height, for twitter bootstrap */
	  line-height: 1em;
	 
	  /* Animation center compensation - margins should be symmetric */
	  /* remove if not needed */
	  margin-left: .2em;
	
	}
	 
	.icon-spin5:before { content: '\e806'; } /* '' */
	.icon-left-open-1:before { content: '\e81e'; } /* '' */
	.icon-plus-1:before { content: '\e801'; } /* '' */
	.icon-facebook:before { content: '\e816'; } /* '' */
	.icon-gplus:before { content: '\e818'; } /* '' */
	.icon-linkedin-3:before { content: '\e815'; } /* '' */
	.icon-twitter:before { content: '\e817'; } /* '' */
	.icon-youtube-play:before { content: '\e805'; } /* '' */
	.icon-plus-2:before { content: '\e809'; } /* '' */
	.icon-record-1:before { content: '\e81c'; } /* '' */
	.icon-cancel:before { content: '\e80b'; } /* '' */
	.icon-search:before { content: '\e804'; } /* '' */
	.icon-cancel-2:before { content: '\e80c'; } /* '' */
	.icon-plus-4:before { content: '\e80d'; } /* '' */
	.icon-cancel-3:before { content: '\e80e'; } /* '' */
	.icon-plus:before { content: '\e808'; } /* '' */
	.icon-down-open:before { content: '\e800'; } /* '' */
	.icon-right-open:before { content: '\e802'; } /* '' */
	.icon-up-open:before { content: '\e803'; } /* '' */
	.icon-stop:before { content: '\e813'; } /* '' */
	.icon-cancel-1:before { content: '\e820'; } /* '' */
	.icon-plus-3:before { content: '\e80a'; } /* '' */
	/* end icon font helpers */

	/* icon font spin animation */
	.animate-spin {
	  -moz-animation: spin 2s infinite linear;
	  -o-animation: spin 2s infinite linear;
	  -webkit-animation: spin 2s infinite linear;
	  animation: spin 2s infinite linear;
	  display: inline-block;
	}
	@-moz-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-webkit-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-o-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-ms-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	/*  end spin animation */	

	@font-face {
	    font-family: 'bebas_neueregular';
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot');
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot?#iefix') format('embedded-opentype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.woff') format('woff'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.ttf') format('truetype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.svg#bebas_neueregular') format('svg');
	    font-weight: normal;
	    font-style: normal;
	}

	/* top chrome styles */
	#top_chrome {

		width:100%;
		height:28px;
		background-color:#163E65;
	}

	.chrome_link_block {

		color:#fff !important;	
		display:inline-block;
		height:100%;
		max-height:28px;
	}

	.chrome_link_block:hover {

		background-color:#eee;
		color:#0779BE !important;	
		cursor:pointer;
	}

	#top_chrome a {

		color:inherit !important;	
		text-decoration:none;
	}

	#top_chrome a:hover {

		color:inherit !important;	
	}

	#top_chrome_buttons {

		float:right;
		margin-right:50px;
	}

	.chrome_button {

		font-family:arial,sans-serif;
		font-weight:bold;
		display:inline-block;
		font-size:9px;
		text-align:left;
		text-decoration:none;
		margin:9px 20px 10px 20px;
	}

	.chrome_active {

		background-color:#1b7abd;
		color:#fff;
	}


	/* header element styles*/
	#header_content {

		font-family: bebas_neueregular, arial, sans-serif;
		font-size:22px;
		color:#0779be;		
		letter-spacing:0.03em;

		position:relative;
		width:calc(100% - 112px);
		min-width:800px;
		height:76px;
		border-bottom:2px solid #A5A8AB;
		margin-left:50px;
		margin-right:50px;
        z-index:9999; /* IT-9508 */
	}

	#portal_logo {

		position:absolute;
		top:20px;
		left:0px;
		width:236px;
	}

	#portal_tabs {

		position: absolute;
		bottom: 12px;
		left: 0px;
		margin-bottom: -12px;
		display: block;
	}

	.horizontal_list
	{
		margin: 0;
		padding: 0;
		list-style-type: none;
		display: inline; 		
	}

	.horizontal_list li { 

		display: inline; 	
	}	

	#portal_tab_menu_container {

		display:none;
	}

	#portal_search_input_group {

		position:absolute;
		margin: 0;
		padding: 0;
		bottom:12px;
		right:0px;
		border:1px solid #DAD9D3;
		color:#DAD9D3;
	}

	#portal_search_input_group input {

		margin: 0;
		padding: 0;
		border:none;
		width:110px;
		color:#DAD9D3;
		font-family:Helvetica,sans-serif;
		font-size:12px;
		height:28px;
		margin-left:10px;

		outline:none;
	}

	input::-ms-clear {
		display: none;
	}

	#portal_search_input_group i {

		font-size:0.8em;
		cursor:pointer;
	}

	#pSiteSearchTextInput:focus {

		color:#000;
	}

	.icon_menu_down {

		font-size: 12px;
		margin-left: 4px;
		vertical-align: text-top;
		padding-top: 4px;
		display: inline-block;
	}

	.portal_tab_item {

		background-color: #fff;
		padding: 12px;
		padding-left:20px;
		padding-right:20px;
		display: inline-block !important;		
	}

	.portal_tab_item a:hover {

		color:#fbe300;	
	}

	.portal_tab_item:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}	


	.portal_menu_active {

		background-color: #EEEEEE;
	}	

	.portal_menu_item_active {

		color:#fbe300 !important;	
	}

	.portal_tab_active {

		color:#a5a8ab !important;
	}

	#header_content a {		

		color:inherit !important;
	}

	#header_content a:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}

	#header_content ul {

		list-style: none;
		position: relative;
		top: 58px;
		display: inline-table;
	}

	#header_content ul ul {
		
		display:none;
		opacity:0;
	}

	/* the fly out menus */
	#header_content ul li:hover > ul,
	#header_content ul li.hover > ul {

		position: absolute;
		top: 48px;
		z-index:999;

		display: block;
		width: 152px;
		padding:0px;
		padding-top:20px;
		padding-bottom:10px;
		font-size: 12px;
		font-family: Arial,Sans-Serif;
		font-weight: bold;
		text-transform: uppercase;

		background-color: rgba(8,122,191,0.98);
		color: #fff !important;
		opacity: 1;

		-webkit-animation: fade 0.25s;
		animation: fade 0.25s;
	}

	@-webkit-keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}
	@keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}	

	#header_content ul li:hover > ul li {

		padding: 10px;
		padding-top: 0px;
		padding-left: 20px;
		display: block;
		opacity:1;
	}

	#header_content ul li:hover > ul a {

		display: block;
		opacity:1;
		white-space:nowrap;
	}

	#portal_menu_start_list {

		left: 81px;

		/* IE sauce */
		left:85px\9;

		/* Safari sauce */
		(;left:79px;);
	}

	#portal_menu_atlases_list {

		left: 290px;
		width: 210px !important;
        height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 303px\9;
        height: 225px\9;

		/* Safari sauce */
		(;left:279px;);
	}

	#portal_menu_tools_list {

		left: 500px;
		width: 172px !important;
		height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 513px\9;
		height:225px\9;

		/* Safari sauce */
		(;left:489px;);
		(;height:227px;);
	}

	/* mozilla sauce */
	@-moz-document url-prefix() {

		#header_content {

			height:107px;
		}

		#portal_menu_start_list {

			left:85px;
		}

		#portal_menu_atlases_list {

			left: 302px;		
		}

		#portal_menu_tools_list {

			left: 512px;
			height:236px;
		}
	}


	/* make it look OK on iPad also */
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:portrait) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}  
	}
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:landscape) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}	  
	}

	.portal_menu_list_heading {

		color:#9ccae5;
		border-bottom:1px solid #9ccae5;
		padding-bottom:5px !important;
	}

	.portal_menu_list_heading + li {

		padding-top:10px !important;
	}

	.portal_menu_list_heading li {

		padding-bottom:0px;
	}	

	#portal_menu_atlases_list li:last-of-type {

		padding-bottom:15px !important;
	}

	.icon-spin5 {

		color:#000;
	}
</style>
<div id="pHeader">
<div id="top_chrome">
	<div id="top_chrome_buttons">
		<div class="chrome_link_block"><a href="http://www.alleninstitute.org" target="_blank" class="chrome_button">ALLEN INSTITUTE</a></div>
	</div>
</div>

<div id="header_content">

	<a href="http://AllenInstitute.github.io">
		<img id="portal_logo" style="border:0;" src="_static/external_assets/images/AIBS_Logo.png"/>
	</a>
</div>
</div>

<script>

jQuery('#portal_tab_menu').bind('touchstart', function() {

	if(window.console)
		console.log("touchstart");
});
</script>
<link rel="stylesheet" type="text/css" href="_static/external_assets/stylesheets/common_layout.css" />


  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Tutorial-1:-Single-cell-simulation-with-current-injection-(with-BioNet)">
<h1>Tutorial 1: Single cell simulation with current injection (with BioNet)<a class="headerlink" href="#Tutorial-1:-Single-cell-simulation-with-current-injection-(with-BioNet)" title="Permalink to this headline">¶</a></h1>
<p>In this example we will build a network consisting of a single biophysically detailed cell. Then we will run a short simulation with a current injection at the soma of the cell, then look at the output of spikes, membrane potential and calcium flux.</p>
<p><strong>Note</strong> - scripts and files for running this tutorial can be found in the directory <a class="reference external" href="https://github.com/AllenInstitute/bmtk/tree/develop/docs/tutorial/sources/chapter01">sources/chapter01/</a></p>
<p><strong>requirements:</strong> * Python 2.7, 3.6+ * bmtk * NEURON 7.4+</p>
<section id="1.-Building-the-network">
<h2>1. Building the network<a class="headerlink" href="#1.-Building-the-network" title="Permalink to this headline">¶</a></h2>
<p>The first step is to use the bmtk Network Builder to create and save the network. First we instantiate a network with a name or our choosing (since throughout this tutorial we will use cell models from the mouse cortex, let’s call our network ‘mcortex’).</p>
<p>Once we have initialized a network object, we can add a single node by calling the add_nodes method.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.networks</span> <span class="kn">import</span> <span class="n">NetworkBuilder</span>

<span class="n">net</span> <span class="o">=</span> <span class="n">NetworkBuilder</span><span class="p">(</span><span class="s1">&#39;mcortex&#39;</span><span class="p">)</span>
<span class="n">net</span><span class="o">.</span><span class="n">add_nodes</span><span class="p">(</span>
    <span class="n">cell_name</span><span class="o">=</span><span class="s1">&#39;Scnn1a_473845048&#39;</span><span class="p">,</span>
    <span class="n">potential</span><span class="o">=</span><span class="s1">&#39;exc&#39;</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s1">&#39;biophysical&#39;</span><span class="p">,</span>
    <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;ctdb:Biophys1.hoc&#39;</span><span class="p">,</span>
    <span class="n">model_processing</span><span class="o">=</span><span class="s1">&#39;aibs_perisomatic&#39;</span><span class="p">,</span>
    <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;472363762_fit.json&#39;</span><span class="p">,</span>
    <span class="n">morphology</span><span class="o">=</span><span class="s1">&#39;Scnn1a_473845048_m.swc&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Some of the parameters used to create the node are optional and only for our benefit. Others are necessary for when we will eventually run a simulation: * <em>cell_name</em> (optional) - Name/type of cell we will be modeling. * <em>potential</em> (optional) - Use to indicate that it is an excitatory type cell. * <em>model_type</em> - Used by the simulator to indicate that we are using a biophysical cell. * <em>model template</em> - NEURON .hoc file, NeuroML file, or other template used in the creation of cell/node
object. * <em>model_processing</em> - A custom function used by the simulator to load the model into NEURON using Allen Cell Types files for perisomatic models. * <em>dynamics_params</em> - Model parameters. File will be downloaded from the Allen Cell Types Database. * <em>morphology</em> - Model morphology. File will be downloaded from the Allen Cell Types Database.</p>
<section id="Building-and-saving">
<h3>Building and saving<a class="headerlink" href="#Building-and-saving" title="Permalink to this headline">¶</a></h3>
<p>The final thing to do is to build and save the network. If successful, we should see a combination of hdf5 and csv files in the ‘./network’ directory, these files are used describe the network, and can be saved, stored and run at a later date.</p>
<p>First, it’s a good idea to remove any old files in the “network” folder so they don’t interfere with the current simulation.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ rm -r sim_ch01/network
</pre></div>
</div>
<p>If you get the output:</p>
<p><code class="docutils literal notranslate"><span class="pre">rm:</span> <span class="pre">cannot</span> <span class="pre">remove</span> <span class="pre">sim_ch01/network/*:</span> <span class="pre">No</span> <span class="pre">such</span> <span class="pre">file</span> <span class="pre">or</span> <span class="pre">directory</span></code></p>
<p>It’s OK. Keep going.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">net</span><span class="o">.</span><span class="n">build</span><span class="p">()</span>
<span class="n">net</span><span class="o">.</span><span class="n">save_nodes</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch01/network&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Use the NetworkBuilder nodes() method to show that a node of our parameters was created</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">net</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;cell_name&#39;: &#39;Scnn1a_473845048&#39;, &#39;potential&#39;: &#39;exc&#39;, &#39;model_type&#39;: &#39;biophysical&#39;, &#39;model_template&#39;: &#39;ctdb:Biophys1.hoc&#39;, &#39;model_processing&#39;: &#39;aibs_perisomatic&#39;, &#39;dynamics_params&#39;: &#39;472363762_fit.json&#39;, &#39;morphology&#39;: &#39;Scnn1a_473845048_m.swc&#39;, &#39;node_type_id&#39;: 100, &#39;node_id&#39;: 0}
</pre></div></div>
</div>
</section>
</section>
<section id="2.-Setting-up-the-simulator-environment">
<h2>2. Setting up the simulator environment<a class="headerlink" href="#2.-Setting-up-the-simulator-environment" title="Permalink to this headline">¶</a></h2>
<p>Now that the network has been built, we can use the BioNet simulator to setup and run it using NEURON. The easiest ways to do this is to copy and modify <a class="reference external" href="https://github.com/AllenInstitute/bmtk/tree/develop/docs/examples/bio_14cells">existing simulation setup</a>, or use the <code class="docutils literal notranslate"><span class="pre">bmtk.utils.sim_setup</span></code> script which can be called from the command-line</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m bmtk.utils.sim_setup  --report-vars v,cai           <span class="se">\</span>
                                  --network sim_ch01/network/   <span class="se">\</span>
                                  --iclamp <span class="m">0</span>.12,500.0,1000.0    <span class="se">\</span>
                                  --dt <span class="m">0</span>.1 --tstop <span class="m">2000</span>.0       <span class="se">\</span>
                                  --include-examples            <span class="se">\</span>
                                  --compile-mechanisms          <span class="se">\</span>
                                  bionet sim_ch01
</pre></div>
</div>
<p>Or call the function directly in Python:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.utils.sim_setup</span> <span class="kn">import</span> <span class="n">build_env_bionet</span>

<span class="n">build_env_bionet</span><span class="p">(</span>
    <span class="n">base_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch01&#39;</span><span class="p">,</span>       <span class="c1"># Where to save the scripts and config files</span>
    <span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;config.json&#39;</span><span class="p">,</span> <span class="c1"># Where main config will be saved.</span>
    <span class="n">network_dir</span><span class="o">=</span><span class="s1">&#39;network&#39;</span><span class="p">,</span>     <span class="c1"># Location of directory containing network files</span>
    <span class="n">tstop</span><span class="o">=</span><span class="mf">2000.0</span><span class="p">,</span> <span class="n">dt</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>      <span class="c1"># Run a simulation for 2000 ms at 0.1 ms intervals</span>
    <span class="n">report_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;cai&#39;</span><span class="p">],</span>  <span class="c1"># Tells simulator we want to record membrane potential and calcium traces</span>
    <span class="n">current_clamp</span><span class="o">=</span><span class="p">{</span>            <span class="c1"># Creates a step current from 500.0 ms to 1500.0 ms</span>
        <span class="s1">&#39;amp&#39;</span><span class="p">:</span> <span class="mf">0.120</span><span class="p">,</span>
        <span class="s1">&#39;delay&#39;</span><span class="p">:</span> <span class="mf">500.0</span><span class="p">,</span>
        <span class="s1">&#39;duration&#39;</span><span class="p">:</span> <span class="mf">1000.0</span>
    <span class="p">},</span>
    <span class="n">include_examples</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># Copies components files for tutorial examples</span>
    <span class="n">compile_mechanisms</span><span class="o">=</span><span class="kc">True</span>   <span class="c1"># Will try to compile NEURON mechanisms</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:root:Configuration file /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/config.json already exists, skipping. Please delete existing file, use a different name, or use overwrite_config=True.
WARNING:root:Configuration file /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/circuit_config.json already exists, skipping. Please delete existing file, use a different name, or use overwrite_config=True.
WARNING:root:Configuration file /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/simulation_config.json already exists, skipping. Please delete existing file, use a different name, or use overwrite_config=True.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
/home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms
Mod files: &#34;modfiles/CaDynamics.mod&#34; &#34;modfiles/Ca_HVA.mod&#34; &#34;modfiles/Ca_LVA.mod&#34; &#34;modfiles/Ih.mod&#34; &#34;modfiles/Im.mod&#34; &#34;modfiles/Im_v2.mod&#34; &#34;modfiles/Kd.mod&#34; &#34;modfiles/K_P.mod&#34; &#34;modfiles/K_T.mod&#34; &#34;modfiles/Kv2like.mod&#34; &#34;modfiles/Kv3_1.mod&#34; &#34;modfiles/Nap.mod&#34; &#34;modfiles/NaTa.mod&#34; &#34;modfiles/NaTs.mod&#34; &#34;modfiles/NaV.mod&#34; &#34;modfiles/SK.mod&#34; &#34;modfiles/vecevent.mod&#34;

COBJS=&#39;&#39;
 -&gt; <span class="ansi-green-fg">Compiling</span> mod_func.cpp
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/CaDynamics.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Ca_HVA.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Ca_LVA.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Ih.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Im.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Im_v2.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Kd.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/K_P.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/K_T.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Kv2like.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Kv3_1.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/Nap.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/NaTa.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/NaTs.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/NaV.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/SK.mod
 -&gt; <span class="ansi-green-fg">NMODL</span> ../modfiles/vecevent.mod
 -&gt; <span class="ansi-green-fg">Compiling</span> CaDynamics.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Ca_HVA.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Ca_LVA.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Ih.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Im.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Im_v2.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Kd.c
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Translating Ca_HVA.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Ca_HVA.c
Translating Ca_LVA.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Ca_LVA.c
Translating CaDynamics.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/CaDynamics.c
Thread Safe
Thread Safe
Thread Safe
Translating Ih.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Ih.c
Translating Im.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Im.c
Translating Im_v2.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Im_v2.c
Thread Safe
Thread Safe
Thread Safe
Translating Kd.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Kd.c
Translating K_P.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/K_P.c
Translating K_T.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/K_T.c
Thread Safe
Thread Safe
Thread Safe
Translating Kv2like.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Kv2like.c
Translating Kv3_1.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Kv3_1.c
Translating Nap.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/Nap.c
Thread Safe
Thread Safe
Thread Safe
Translating NaTa.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/NaTa.c
Translating NaTs.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/NaTs.c
Translating NaV.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/NaV.c
Thread Safe
Thread Safe
NEURON&#39;s CVode method ignores conservation
Notice: LINEAR is not thread safe.
Translating SK.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/SK.c
Translating vecevent.mod into /home/ping/bmtk_change/bmtk/docs/tutorial/sim_ch01/components/mechanisms/x86_64/vecevent.c
Notice: VERBATIM blocks are not thread safe
Thread Safe
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 -&gt; <span class="ansi-green-fg">Compiling</span> K_P.c
 -&gt; <span class="ansi-green-fg">Compiling</span> K_T.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Kv2like.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Kv3_1.c
 -&gt; <span class="ansi-green-fg">Compiling</span> Nap.c
 -&gt; <span class="ansi-green-fg">Compiling</span> NaTa.c
 -&gt; <span class="ansi-green-fg">Compiling</span> NaTs.c
 -&gt; <span class="ansi-green-fg">Compiling</span> NaV.c
 -&gt; <span class="ansi-green-fg">Compiling</span> SK.c
 -&gt; <span class="ansi-green-fg">Compiling</span> vecevent.c
 =&gt; <span class="ansi-green-fg">LINKING</span> shared library ./libnrnmech.so
Successfully created x86_64/special
</pre></div></div>
</div>
<div class="admonition warning">
<p class="admonition-title"><strong>Warning</strong>: You may see an error that it failed while trying to compile the NEURON mechanisms - especially if Jupyter is unable to find your NEURON environment path. If this happens you will need to manually compile the mechanisms:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>cd sim_ch01/components/mechanisms
nrnivmodl modfiles
</pre></div>
</div>
</div>
<p>This will create the directory <strong>sim_ch01</strong> (feel free to give it a better name or set it to a different locations) with a number of new files and folders. Some of the more important ones include: * <strong>circuit_config.json</strong> - A configuration file that contains the location of the network files we created above, and the location of neuron and synaptic models, templates, morphologies and mechanisms required to build and instantiate individual cell models.</p>
<ul class="simple">
<li><p><strong>simulation_config.json</strong> - contains information about the simulation. Including initial conditions and run-time configuration (<em>run</em> and <em>conditions</em>). In the <em>inputs</em> section we define what external sources we will use to drive the network (in this case a current clamp). And in the <em>reports</em> section we define the variables (somatic membrane potential and calcium level) that will be recorded during the simulation.</p></li>
<li><p><strong>components/biophysical_neuron_models/472363762_fit.json</strong> - The parameter file for the cell we’re modeling. Originally <a class="reference external" href="http://celltypes.brain-map.org/neuronal_model/download/482934212">downloaded from the Allen Cell Types Database</a></p></li>
<li><p><strong>components/biophysical_neuron_models/Scnn1a_473845048_m.swc</strong> - The morphology file for our cell. Originally <a class="reference external" href="http://celltypes.brain-map.org/neuronal_model/download/482934212">downloaded from the Allen Cell Types Database</a></p></li>
</ul>
<p>Modifying these files with a text editor, or swapping out different parameter or morphologies, allows for changing the simulation without extra programming.</p>
</section>
<section id="3.-Running-the-simulation">
<h2>3. Running the simulation<a class="headerlink" href="#3.-Running-the-simulation" title="Permalink to this headline">¶</a></h2>
<p>Once our config file is setup we can run a simulation either through the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">cd</span> sim_ch01
$ python run_bionet.py config.json
</pre></div>
</div>
<p>or through the script</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.simulator</span> <span class="kn">import</span> <span class="n">bionet</span>

<span class="n">conf</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">Config</span><span class="o">.</span><span class="n">from_json</span><span class="p">(</span><span class="s1">&#39;sim_ch01/config.json&#39;</span><span class="p">)</span>
<span class="n">conf</span><span class="o">.</span><span class="n">build_env</span><span class="p">()</span>
<span class="n">conf</span>
<span class="n">net</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">BioNetwork</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">)</span>
<span class="n">sim</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">BioSimulator</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">,</span> <span class="n">network</span><span class="o">=</span><span class="n">net</span><span class="p">)</span>
<span class="n">sim</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,159 [INFO] Created log file
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Created log file
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
numprocs=1
2022-07-28 12:16:11,226 [INFO] Building cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Building cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,393 [INFO] Building recurrent connections
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Building recurrent connections
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,402 [INFO] Running simulation for 2000.000 ms with the time step 0.100 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Running simulation for 2000.000 ms with the time step 0.100 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,403 [INFO] Starting timestep: 0 at t_sim: 0.000 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Starting timestep: 0 at t_sim: 0.000 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,404 [INFO] Block save every 5000 steps
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Block save every 5000 steps
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:11,692 [INFO]     step:5000 t_sim:500.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:    step:5000 t_sim:500.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:12,006 [INFO]     step:10000 t_sim:1000.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:    step:10000 t_sim:1000.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:12,299 [INFO]     step:15000 t_sim:1500.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:    step:15000 t_sim:1500.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:12,603 [INFO]     step:20000 t_sim:2000.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:    step:20000 t_sim:2000.00 ms
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2022-07-28 12:16:12,627 [INFO] Simulation completed in 1.225 seconds
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:NEURONIOUtils:Simulation completed in 1.225 seconds
</pre></div></div>
</div>
<div class="admonition warning">
<p class="admonition-title"><strong>Warning</strong>: If you get the following error argument not a density mechanism name, you will need to compile the NEURON mechanisms</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>cd sim_ch01/components/mechanisms
nrnivmodl modfiles
</pre></div>
</div>
</div>
<p>A quick breakdown of the script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">conf</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">from_json</span><span class="p">(</span><span class="s1">&#39;config.json&#39;</span><span class="p">)</span>
<span class="n">io</span><span class="o">.</span><span class="n">setup_output_dir</span><span class="p">(</span><span class="n">conf</span><span class="p">)</span>
<span class="n">nrn</span><span class="o">.</span><span class="n">load_neuron_modules</span><span class="p">(</span><span class="n">conf</span><span class="p">)</span>
</pre></div>
</div>
<p>This section loads the configuration file, it sets up the output directory and files for writing during the simulation, and loads NEURON mechanisms needed by the cell model(s) during the simulation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">net</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">BioNetwork</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">)</span>
</pre></div>
</div>
<p>Creates a NEURON representation of the network, including cell models that have been converted into their NEURON equivalents.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sim</span> <span class="o">=</span> <span class="n">Simulation</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">,</span> <span class="n">network</span><span class="o">=</span><span class="n">net</span><span class="p">)</span>
<span class="n">sim</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
<p>Sets up and runs the NEURON simulation. When finished, the outputs (e.g., spike times, membrane potential, and Ca2+ concentration) will be saved into the output directory as specified in the config.</p>
</section>
<section id="4.-Analyzing-the-run">
<h2>4. Analyzing the run<a class="headerlink" href="#4.-Analyzing-the-run" title="Permalink to this headline">¶</a></h2>
<p>The results of the simulation are placed into various files as specified in the “output” section of the config file. We can change this before run-time if required.</p>
<p>All simulations will save the spike times of the network cells. These are saved in csv format (output/spikes.txt) or hdf5 format(output/spikes.h5). To get the table of spike times for our single-cell network we can run the following method from the analyzer (node_id 0 corresponds to our single cell).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.analyzer.spike_trains</span> <span class="kn">import</span> <span class="n">to_dataframe</span>

<span class="n">to_dataframe</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;sim_ch01/config.json&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>timestamps</th>
      <th>node_ids</th>
      <th>population</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>554.8</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1379.5</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1319.5</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1259.5</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1199.5</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>5</th>
      <td>1139.7</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>6</th>
      <td>1079.9</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>7</th>
      <td>1020.4</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>8</th>
      <td>961.2</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>9</th>
      <td>902.6</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>10</th>
      <td>844.9</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>11</th>
      <td>788.7</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>12</th>
      <td>734.7</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>13</th>
      <td>683.9</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>14</th>
      <td>637.1</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>15</th>
      <td>594.4</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>16</th>
      <td>1439.6</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
    <tr>
      <th>17</th>
      <td>1499.7</td>
      <td>0</td>
      <td>mcortex</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>When setting up the environment and config file we specified cell_vars=[‘v’, ‘cai’]. This indicates to the simulator to also record membrane potential and Ca2+ level (and we can also specify other variables as long as they are supported in NEURON).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.analyzer.compartment</span> <span class="kn">import</span> <span class="n">plot_traces</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">plot_traces</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;sim_ch01/config.json&#39;</span><span class="p">,</span> <span class="n">node_ids</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">report_name</span><span class="o">=</span><span class="s1">&#39;v_report&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">plot_traces</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;sim_ch01/config.json&#39;</span><span class="p">,</span> <span class="n">node_ids</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">report_name</span><span class="o">=</span><span class="s1">&#39;cai_report&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_single_cell_clamped_16_0.png" src="_images/tutorial_single_cell_clamped_16_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_single_cell_clamped_16_1.png" src="_images/tutorial_single_cell_clamped_16_1.png" />
</div>
</div>
</section>
<section id="5.-Additional-Information">
<h2>5. Additional Information<a class="headerlink" href="#5.-Additional-Information" title="Permalink to this headline">¶</a></h2>
<section id="Changing-run-time-parameters.">
<h3>Changing run-time parameters.<a class="headerlink" href="#Changing-run-time-parameters." title="Permalink to this headline">¶</a></h3>
<p>By making changes to the config file, we can change the conditions and simulation parameters without having to rebuild the network, modify parameter files, or change our run_bionet script. In fact we can iteratively run multiple simulations without any extra coding, only a text editor to change the json file.</p>
<p>The run section of the config.json contains most of the parameters unique to the simulation:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="nt">&quot;run&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;dL&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">20</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;nsteps_block&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">5000</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;spike_threshold&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">-15</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;tstop&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">2000.0</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;dt&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.1</span><span class="w"></span>
<span class="w">  </span><span class="p">}</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
<ul class="simple">
<li><p><em>tstop</em> - Simulation runtime in milliseconds.</p></li>
<li><p><em>dt</em> - Time steps of the simulation. Decreasing dt should increase accuracy of firing dynamics, but also increase time it takes to complete.</p></li>
<li><p><em>spike_threshold</em> - Used to determine when to count an action potential</p></li>
<li><p><em>dL</em> - Length of segments in a section. The number of segments will be forced to an odd number so the actual length may vary.</p></li>
</ul>
<p>Through the conditions section we can adjust simulation temperature (C) and the initial membrane potential of the cells:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="nt">&quot;conditions&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;celsius&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">34.0</span><span class="p">,</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;v_init&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">-80</span><span class="w"></span>
<span class="w">  </span><span class="p">}</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
<p>And lastly, the input section lets us control stimulus onto the network. There are a number of different options which will be explained in the following tutorials. But even with a simple current injection we can adjust amplitude, delay and stimulation duration and measure the effect on the cell.</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="nt">&quot;inputs&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;current_clamp&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;input_type&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;current_clamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;module&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;IClamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;node_set&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;all&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;amp&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.120</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;delay&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">500.0</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;duration&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">1000.0</span><span class="w"></span>
<span class="w">    </span><span class="p">}</span><span class="w"></span>
<span class="w">  </span><span class="p">}</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
<p>We can even add multiple injections</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span><span class="w"></span>
<span class="w">  </span><span class="nt">&quot;inputs&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">    </span><span class="nt">&quot;cclamp1&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;input_type&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;current_clamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;module&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;IClamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;node_set&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;all&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;amp&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.150</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;delay&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.0</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;duration&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">400.0</span><span class="w"></span>
<span class="w">    </span><span class="p">},</span><span class="w"></span>

<span class="w">    </span><span class="nt">&quot;cclamp2&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;input_type&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;current_clamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;module&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;IClamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;node_set&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;all&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;amp&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.300</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;delay&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">500.0</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;duration&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">400.0</span><span class="w"></span>
<span class="w">    </span><span class="p">},</span><span class="w"></span>

<span class="w">    </span><span class="nt">&quot;cclamp3&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;input_type&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;current_clamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;module&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;IClamp&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;node_set&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;all&quot;</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;amp&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.450</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;delay&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">1000.0</span><span class="p">,</span><span class="w"></span>
<span class="w">      </span><span class="nt">&quot;duration&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">400.0</span><span class="w"></span>
<span class="w">    </span><span class="p">}</span><span class="w"></span>
<span class="w">  </span><span class="p">}</span><span class="w"></span>
<span class="p">}</span><span class="w"></span>
</pre></div>
</div>
</section>
<section id="Changing-cell-models">
<h3>Changing cell models<a class="headerlink" href="#Changing-cell-models" title="Permalink to this headline">¶</a></h3>
<p>When building the network we defined the cell model and morphology through the ‘dynamics_params’ and ‘morphology_file’ options. After building and saving the network, these values were saved in the node-types csv file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;sim_ch01/network/mcortex_node_types.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>node_type_id</th>
      <th>model_template</th>
      <th>morphology</th>
      <th>model_processing</th>
      <th>potential</th>
      <th>model_type</th>
      <th>dynamics_params</th>
      <th>cell_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>100</td>
      <td>ctdb:Biophys1.hoc</td>
      <td>Scnn1a_473845048_m.swc</td>
      <td>aibs_perisomatic</td>
      <td>exc</td>
      <td>biophysical</td>
      <td>472363762_fit.json</td>
      <td>Scnn1a_473845048</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>If we want to run the simulation on a different cell model, all we have to do is: 1. Download new parameters.json and morphology.swc into components/biophysical_neuron_models and components/morphologies 2. Open mcortex_node_types.csv in a text editor and update ‘morphology_file’ and ‘params_file’ accordingly.</p>
<p>In our simple one-cell example, it is likely faster to just rebuild the network. However the advantage of the use of the node types becomes clear once we start dealing with a larger network. For example we may have a network of hundreds of thousands of individual cells with tens of thousands of Scnn1a type cells. The process of adjusting/changing the Scnn1a parameter in the csv then starting another simulation only takes seconds, whereas rebuilding the entire network may take hours.</p>
</section>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h3><a href="index.html">Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="builder.html">Building Networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulators.html">Simulation Engines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bionet.html">BioNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="pointnet.html">PointNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="popnet.html">PopNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="filternet.html">FilterNet</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analyzer.html">Analyzing the Results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_NetworkBuilder_Intro.html">Builder: Using the Network Builder</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">BioNet: Single cell with current injection</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_cell_syn.html">BioNet: Single with with synaptic input</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_pop.html">BioNet: Multiple Nodes with single cell-type</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_multi_pop.html">BioNet: Heterogeneous network</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_pointnet_modeling.html">PointNet: Point-neuron modeling</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_population_modeling.html">PopNet: Population-based firing rate models</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_filter_models.html">FilterNet: Full-field flashing movie</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Example Networks</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/AllenInstitute/bmtk.git">Github Profile</a></li>
<li class="toctree-l1"><a class="reference internal" href="workshop.html">Annual Modeling Software Workshop</a></li>
<li class="toctree-l1"><a class="reference internal" href="registration.html">Registration</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributors.html">Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_cite.html">How to Cite</a></li>
<li class="toctree-l1"><a class="reference internal" href="bmtk/bmtk.html">Source Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.analyzer.html">bmtk.analyzer package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.builder.html">bmtk.builder package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.simulator.html">bmtk.simulator package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.tests.html">bmtk.tests package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.utils.html">bmtk.utils package</a></li>
</ul>
</li>
</ul>

<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


    <div class="footer" role="contentinfo">
    </div>
<script type="text/javascript" src="_static/external_assets/javascript/portalFooter.js"></script>

  </body>
</html>