
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Tools for generating cell placements &#8212; Brain Modeling Toolkit 1.0.7 documentation</title>
    <link rel="stylesheet" href="_static/aibs_sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Synaptic plasticity in Pointnet (STP, facilitation, depression, STDP, and others)" href="tutorial_dynamic_synapses.html" />
    <link rel="prev" title="Additional Tutorials" href="ad_tutorials.html" /> 
  </head><body>
<script type="text/javascript" src="_static/external_assets/javascript/portal.js"></script>
<style>

	/* the icon font-family definition and helper classes */
	@font-face {
	  font-family: 'portal_icon_font';
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677');
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677#iefix') format('embedded-opentype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.woff?39505677') format('woff'),
	       url('_static/external_assets/stylesheets/portal_icon_font.ttf?39505677') format('truetype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.svg?39505677#portal_icon_font') format('svg');
	  font-weight: normal;
	  font-style: normal;
	}
	 
	 [class^="icon-"]:before, [class*=" icon-"]:before {
	  font-family: "portal_icon_font";
	  font-style: normal;
	  font-weight: normal;
	  speak: none;
	 
	  display: inline-block;
	  text-decoration: inherit;
	  width: 1em;
	  margin-right: .2em;
	  text-align: center;
	  /* opacity: .8; */
	 
	  /* For safety - reset parent styles, that can break glyph codes*/
	  font-variant: normal;
	  text-transform: none;
	     
	  /* fix buttons height, for twitter bootstrap */
	  line-height: 1em;
	 
	  /* Animation center compensation - margins should be symmetric */
	  /* remove if not needed */
	  margin-left: .2em;
	
	}
	 
	.icon-spin5:before { content: '\e806'; } /* '' */
	.icon-left-open-1:before { content: '\e81e'; } /* '' */
	.icon-plus-1:before { content: '\e801'; } /* '' */
	.icon-facebook:before { content: '\e816'; } /* '' */
	.icon-gplus:before { content: '\e818'; } /* '' */
	.icon-linkedin-3:before { content: '\e815'; } /* '' */
	.icon-twitter:before { content: '\e817'; } /* '' */
	.icon-youtube-play:before { content: '\e805'; } /* '' */
	.icon-plus-2:before { content: '\e809'; } /* '' */
	.icon-record-1:before { content: '\e81c'; } /* '' */
	.icon-cancel:before { content: '\e80b'; } /* '' */
	.icon-search:before { content: '\e804'; } /* '' */
	.icon-cancel-2:before { content: '\e80c'; } /* '' */
	.icon-plus-4:before { content: '\e80d'; } /* '' */
	.icon-cancel-3:before { content: '\e80e'; } /* '' */
	.icon-plus:before { content: '\e808'; } /* '' */
	.icon-down-open:before { content: '\e800'; } /* '' */
	.icon-right-open:before { content: '\e802'; } /* '' */
	.icon-up-open:before { content: '\e803'; } /* '' */
	.icon-stop:before { content: '\e813'; } /* '' */
	.icon-cancel-1:before { content: '\e820'; } /* '' */
	.icon-plus-3:before { content: '\e80a'; } /* '' */
	/* end icon font helpers */

	/* icon font spin animation */
	.animate-spin {
	  -moz-animation: spin 2s infinite linear;
	  -o-animation: spin 2s infinite linear;
	  -webkit-animation: spin 2s infinite linear;
	  animation: spin 2s infinite linear;
	  display: inline-block;
	}
	@-moz-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-webkit-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-o-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-ms-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	/*  end spin animation */	

	@font-face {
	    font-family: 'bebas_neueregular';
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot');
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot?#iefix') format('embedded-opentype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.woff') format('woff'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.ttf') format('truetype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.svg#bebas_neueregular') format('svg');
	    font-weight: normal;
	    font-style: normal;
	}

	/* top chrome styles */
	#top_chrome {

		width:100%;
		height:28px;
		background-color:#163E65;
	}

	.chrome_link_block {

		color:#fff !important;	
		display:inline-block;
		height:100%;
		max-height:28px;
	}

	.chrome_link_block:hover {

		background-color:#eee;
		color:#0779BE !important;	
		cursor:pointer;
	}

	#top_chrome a {

		color:inherit !important;	
		text-decoration:none;
	}

	#top_chrome a:hover {

		color:inherit !important;	
	}

	#top_chrome_buttons {

		float:right;
		margin-right:50px;
	}

	.chrome_button {

		font-family:arial,sans-serif;
		font-weight:bold;
		display:inline-block;
		font-size:9px;
		text-align:left;
		text-decoration:none;
		margin:9px 20px 10px 20px;
	}

	.chrome_active {

		background-color:#1b7abd;
		color:#fff;
	}


	/* header element styles*/
	#header_content {

		font-family: bebas_neueregular, arial, sans-serif;
		font-size:22px;
		color:#0779be;		
		letter-spacing:0.03em;

		position:relative;
		width:calc(100% - 112px);
		min-width:800px;
		height:76px;
		border-bottom:2px solid #A5A8AB;
		margin-left:50px;
		margin-right:50px;
        z-index:9999; /* IT-9508 */
	}

	#portal_logo {

		position:absolute;
		top:20px;
		left:0px;
		width:236px;
	}

	#portal_tabs {

		position: absolute;
		bottom: 12px;
		left: 0px;
		margin-bottom: -12px;
		display: block;
	}

	.horizontal_list
	{
		margin: 0;
		padding: 0;
		list-style-type: none;
		display: inline; 		
	}

	.horizontal_list li { 

		display: inline; 	
	}	

	#portal_tab_menu_container {

		display:none;
	}

	#portal_search_input_group {

		position:absolute;
		margin: 0;
		padding: 0;
		bottom:12px;
		right:0px;
		border:1px solid #DAD9D3;
		color:#DAD9D3;
	}

	#portal_search_input_group input {

		margin: 0;
		padding: 0;
		border:none;
		width:110px;
		color:#DAD9D3;
		font-family:Helvetica,sans-serif;
		font-size:12px;
		height:28px;
		margin-left:10px;

		outline:none;
	}

	input::-ms-clear {
		display: none;
	}

	#portal_search_input_group i {

		font-size:0.8em;
		cursor:pointer;
	}

	#pSiteSearchTextInput:focus {

		color:#000;
	}

	.icon_menu_down {

		font-size: 12px;
		margin-left: 4px;
		vertical-align: text-top;
		padding-top: 4px;
		display: inline-block;
	}

	.portal_tab_item {

		background-color: #fff;
		padding: 12px;
		padding-left:20px;
		padding-right:20px;
		display: inline-block !important;		
	}

	.portal_tab_item a:hover {

		color:#fbe300;	
	}

	.portal_tab_item:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}	


	.portal_menu_active {

		background-color: #EEEEEE;
	}	

	.portal_menu_item_active {

		color:#fbe300 !important;	
	}

	.portal_tab_active {

		color:#a5a8ab !important;
	}

	#header_content a {		

		color:inherit !important;
	}

	#header_content a:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}

	#header_content ul {

		list-style: none;
		position: relative;
		top: 58px;
		display: inline-table;
	}

	#header_content ul ul {
		
		display:none;
		opacity:0;
	}

	/* the fly out menus */
	#header_content ul li:hover > ul,
	#header_content ul li.hover > ul {

		position: absolute;
		top: 48px;
		z-index:999;

		display: block;
		width: 152px;
		padding:0px;
		padding-top:20px;
		padding-bottom:10px;
		font-size: 12px;
		font-family: Arial,Sans-Serif;
		font-weight: bold;
		text-transform: uppercase;

		background-color: rgba(8,122,191,0.98);
		color: #fff !important;
		opacity: 1;

		-webkit-animation: fade 0.25s;
		animation: fade 0.25s;
	}

	@-webkit-keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}
	@keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}	

	#header_content ul li:hover > ul li {

		padding: 10px;
		padding-top: 0px;
		padding-left: 20px;
		display: block;
		opacity:1;
	}

	#header_content ul li:hover > ul a {

		display: block;
		opacity:1;
		white-space:nowrap;
	}

	#portal_menu_start_list {

		left: 81px;

		/* IE sauce */
		left:85px\9;

		/* Safari sauce */
		(;left:79px;);
	}

	#portal_menu_atlases_list {

		left: 290px;
		width: 210px !important;
        height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 303px\9;
        height: 225px\9;

		/* Safari sauce */
		(;left:279px;);
	}

	#portal_menu_tools_list {

		left: 500px;
		width: 172px !important;
		height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 513px\9;
		height:225px\9;

		/* Safari sauce */
		(;left:489px;);
		(;height:227px;);
	}

	/* mozilla sauce */
	@-moz-document url-prefix() {

		#header_content {

			height:107px;
		}

		#portal_menu_start_list {

			left:85px;
		}

		#portal_menu_atlases_list {

			left: 302px;		
		}

		#portal_menu_tools_list {

			left: 512px;
			height:236px;
		}
	}


	/* make it look OK on iPad also */
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:portrait) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}  
	}
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:landscape) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}	  
	}

	.portal_menu_list_heading {

		color:#9ccae5;
		border-bottom:1px solid #9ccae5;
		padding-bottom:5px !important;
	}

	.portal_menu_list_heading + li {

		padding-top:10px !important;
	}

	.portal_menu_list_heading li {

		padding-bottom:0px;
	}	

	#portal_menu_atlases_list li:last-of-type {

		padding-bottom:15px !important;
	}

	.icon-spin5 {

		color:#000;
	}
</style>
<div id="pHeader">
<div id="top_chrome">
	<div id="top_chrome_buttons">
		<div class="chrome_link_block"><a href="http://www.alleninstitute.org" target="_blank" class="chrome_button">ALLEN INSTITUTE</a></div>
	</div>
</div>

<div id="header_content">

	<a href="http://AllenInstitute.github.io">
		<img id="portal_logo" style="border:0;" src="_static/external_assets/images/AIBS_Logo.png"/>
	</a>
</div>
</div>

<script>

jQuery('#portal_tab_menu').bind('touchstart', function() {

	if(window.console)
		console.log("touchstart");
});
</script>
<link rel="stylesheet" type="text/css" href="_static/external_assets/stylesheets/common_layout.css" />


  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Tools-for-generating-cell-placements">
<h1>Tools for generating cell placements<a class="headerlink" href="#Tools-for-generating-cell-placements" title="Permalink to this headline">¶</a></h1>
<p>Although users are always welcomed to place their neuronal somata according to their own custom generated positions, BMTK provides several tools to facilitate this placement, including simple geometries and placement based on geometries or densities from *.nrrd files. Units can be placed with a minimum distance between them (please see second part of this tutorial).</p>
<div class="section" id="Generate-a-cylindrical-column-of-cell-positions">
<h2>Generate a cylindrical column of cell positions<a class="headerlink" href="#Generate-a-cylindrical-column-of-cell-positions" title="Permalink to this headline">¶</a></h2>
<p>This function places random locations uniformly within a cylinder or cylindrical ring. Note that ‘y’ is the vertical axis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.auxi.node_params</span> <span class="kn">import</span> <span class="n">positions_columnar</span>

<span class="n">positions</span> <span class="o">=</span> <span class="n">positions_columnar</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span><span class="o">=</span><span class="mf">300.0</span><span class="p">,</span> <span class="n">min_radius</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">max_radius</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
                        <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_2_0.png" src="_images/tutorial_cell_placement_2_0.png" />
</div>
</div>
<p>We can also use this function to create a cylindrical ring of locations. For instance, in the Allen V1 model, this method was used to introduce a surround of less computationally intensive unit types around the core of biophysically detailed units.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">positions</span> <span class="o">=</span> <span class="n">positions_columnar</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span><span class="o">=</span><span class="mf">300.0</span><span class="p">,</span> <span class="n">min_radius</span><span class="o">=</span><span class="mf">100.0</span><span class="p">,</span> <span class="n">max_radius</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
                        <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_4_0.png" src="_images/tutorial_cell_placement_4_0.png" />
</div>
</div>
</div>
<div class="section" id="Generate-a-rectangular-prism-volume-of-cell-positions">
<h2>Generate a rectangular prism volume of cell positions<a class="headerlink" href="#Generate-a-rectangular-prism-volume-of-cell-positions" title="Permalink to this headline">¶</a></h2>
<p>Alternatively, if a sheet of cells (with some thickness) or other such geometry is desired, the following function can be used.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.auxi.node_params</span> <span class="kn">import</span> <span class="n">positions_rect_prism</span>

<span class="n">positions</span> <span class="o">=</span> <span class="n">positions_rect_prism</span><span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span><span class="o">=</span><span class="mf">20.0</span><span class="p">,</span> <span class="n">x_length</span><span class="o">=</span><span class="mf">100.0</span><span class="p">,</span> <span class="n">z_length</span><span class="o">=</span><span class="mf">100.0</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_6_0.png" src="_images/tutorial_cell_placement_6_0.png" />
</div>
</div>
</div>
<div class="section" id="Generate-an-ovoid-volume-of-cell-positions">
<h2>Generate an ovoid volume of cell positions<a class="headerlink" href="#Generate-an-ovoid-volume-of-cell-positions" title="Permalink to this headline">¶</a></h2>
<p>Lastly, if an ellipsoid is desired, e.g., to simulate a nucleus, one can use the following function:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.auxi.node_params</span> <span class="kn">import</span> <span class="n">positions_ellipsoid</span>

<span class="n">positions</span> <span class="o">=</span> <span class="n">positions_ellipsoid</span> <span class="p">(</span><span class="mi">400</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span><span class="o">=</span><span class="mf">50.0</span><span class="p">,</span> <span class="n">x_length</span><span class="o">=</span><span class="mf">100.0</span><span class="p">,</span> <span class="n">z_length</span><span class="o">=</span><span class="mf">200.0</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_8_0.png" src="_images/tutorial_cell_placement_8_0.png" />
</div>
</div>
</div>
<div class="section" id="Generate-random-locations-according-to-a-*.nrrd-file">
<h2>Generate random locations according to a *.nrrd file<a class="headerlink" href="#Generate-random-locations-according-to-a-*.nrrd-file" title="Permalink to this headline">¶</a></h2>
<p>We may want to place the units according to a file containing 3D location information, such as the <a class="reference external" href="https://help.brain-map.org/display/mouseconnectivity/API#API-DownloadAtlas3-DReferenceModels">Allen Brain Atlas Common Coordinate Frame</a>. You can obtain a structural mask that describes the extent of a structure of interest. Cells will be placed randomly within that structure at a density of max_dens. If the *.nrrd file contains varying cell densities, this function can also be used to place
cells randomly according to that density raster.</p>
<p>You will need to install the <code class="docutils literal notranslate"><span class="pre">pynrrd</span></code> library to use this function.</p>
<p>Note that the units of cell density are in cells/mm3 while positions are in units of microns.</p>
<p>If the structure is bilateral and only one half is desired for modeling, you can specify the axis along which to split it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.auxi.node_params</span> <span class="kn">import</span> <span class="n">positions_nrrd</span>
<span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;structure_721.nrrd&#39;</span>
<span class="n">max_dens</span> <span class="o">=</span> <span class="mi">100000</span>       <span class="c1"># Cells per mm^3</span>

<span class="n">positions</span> <span class="o">=</span> <span class="n">positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
positions: (102437, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_10_1.png" src="_images/tutorial_cell_placement_10_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">positions</span> <span class="o">=</span> <span class="n">positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="p">,</span> <span class="n">split_bilateral</span><span class="o">=</span><span class="s1">&#39;z&#39;</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
positions: (51218, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_11_1.png" src="_images/tutorial_cell_placement_11_1.png" />
</div>
</div>
<p>You can also directly use a 3-dimensional density matrix with the function <code class="docutils literal notranslate"><span class="pre">positions_density_matrix</span></code> to generate cell placements.</p>
</div>
<div class="section" id="Cell-Placements-with-minimum-distance-between-units">
<h2>Cell Placements with minimum distance between units<a class="headerlink" href="#Cell-Placements-with-minimum-distance-between-units" title="Permalink to this headline">¶</a></h2>
<p>The same built-in placement options above are available now with a minimum distance enforced between cells. With the use of the cell placement object, multiple populations can be placed at the same time, but must use the same minimum distance, which is enforced between and within populations.</p>
<p>For a given minimum distance, there is a maximum density above which it will be physically impossible to pack the units. The functions will inform users of the packing limit for random close packing or hexagonal close packing. The default method is progressive sampling (‘prog’), which iteratively replaces units that violate the minimum distance until the final density is achieved. As this can be slow to run when the requested density is near the packing limit, you may set <code class="docutils literal notranslate"><span class="pre">verbose</span></code> to True to
monitor the progress. You may find that in such cases just slightly decreasing the dmin value significantly speeds up the run.</p>
<p>The alternative jittered lattice method (‘lattice’) may be faster at very high density and may allow you to slightly exceed the random close packing limit, but produces an unrealistic geometric regularity in order to achieve those densities.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.auxi.node_params</span> <span class="kn">import</span> <span class="n">CellLocations</span>

<span class="n">filename</span> <span class="o">=</span> <span class="s1">&#39;structure_721.nrrd&#39;</span>

<span class="n">max_dens</span> <span class="o">=</span> <span class="mi">100000</span>

<span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">18.0</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">CCF_orientation</span><span class="o">=</span><span class="kc">True</span>   <span class="c1"># Use Allen Common Coordinate Framework orientation with y-axis as vertical</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="p">,</span> <span class="n">pop_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pyr&#39;</span><span class="p">],</span> <span class="n">split_bilateral</span><span class="o">=</span><span class="s1">&#39;z&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:206311.963
Max density requested:100000.000
5786/51218 cells placed
11837/51218 cells placed
17823/51218 cells placed
23500/51218 cells placed
28438/51218 cells placed
32752/51218 cells placed
36213/51218 cells placed
39168/51218 cells placed
41516/51218 cells placed
43591/51218 cells placed
45160/51218 cells placed
46412/51218 cells placed
47466/51218 cells placed
48349/51218 cells placed
49117/51218 cells placed
49806/51218 cells placed
50433/51218 cells placed
50989/51218 cells placed
51473/51218 cells placed
51218 cells
positions: (51218, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_14_1.png" src="_images/tutorial_cell_placement_14_1.png" />
</div>
</div>
<p>It is preferable to use partitioning (with the <code class="docutils literal notranslate"><span class="pre">pop_names</span></code> and <code class="docutils literal notranslate"><span class="pre">partitions</span></code> keywords) to place units by the same general rule into the same spatial structure. In the example below, the first call to <code class="docutils literal notranslate"><span class="pre">add_positions_nrrd()</span></code> assigns 85% of returned cells to Pyr and 15% to PV. However, you can also add placements iteratively (second call to <code class="docutils literal notranslate"><span class="pre">add_positions_nrrd()</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">18.0</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">CCF_orientation</span><span class="o">=</span><span class="kc">True</span>   <span class="c1"># Use Allen Common Coordinate Framework orientation with y-axis as vertical</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">pop_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span><span class="s1">&#39;PV&#39;</span><span class="p">],</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">],</span> <span class="n">split_bilateral</span><span class="o">=</span><span class="s1">&#39;z&#39;</span><span class="p">,</span>
                      <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">pop_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pyr2&#39;</span><span class="p">],</span> <span class="n">split_bilateral</span><span class="o">=</span><span class="s1">&#39;z&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:206311.963
Max density requested:50000.000
positions: (25609, 3)
Density limit:206311.963
Max density requested:50000.000
positions: (25609, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_16_1.png" src="_images/tutorial_cell_placement_16_1.png" />
</div>
</div>
<p>Again, you can slightly exceed the above packing limit or speed up placement at densities close to the packing limit by using a jittered lattice method. However, be aware that the resulting placements will have an artifactual regularity.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">20.0</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">CCF_orientation</span><span class="o">=</span><span class="kc">True</span>   <span class="c1"># Use Allen Common Coordinate Framework orientation with y-axis as vertical</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_nrrd</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">max_dens</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">pop_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span><span class="s1">&#39;PV&#39;</span><span class="p">],</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">],</span> <span class="n">split_bilateral</span><span class="o">=</span><span class="s1">&#39;z&#39;</span><span class="p">,</span>
                      <span class="n">method</span><span class="o">=</span><span class="s1">&#39;lattice&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:176661.987
Max density requested:49999.268
positions: (25609, 3)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_18_1.png" src="_images/tutorial_cell_placement_18_1.png" />
</div>
</div>
<p>Here are some more examples with simple geometries, not necessarily all realistic!</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">N</span><span class="o">=</span><span class="mi">4000</span>

<span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;ctx&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">18.0</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_columnar</span><span class="p">([</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span><span class="s1">&#39;PV&#39;</span><span class="p">],</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span>
                               <span class="n">min_radius</span> <span class="o">=</span> <span class="mf">50.0</span><span class="p">,</span> <span class="n">max_radius</span> <span class="o">=</span> <span class="mf">200.0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:206311.963
Max density requested:113176.848
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_20_1.png" src="_images/tutorial_cell_placement_20_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">N</span><span class="o">=</span><span class="mi">2500</span>

<span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">18.0</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_rect_prism</span><span class="p">([</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span><span class="s1">&#39;PV&#39;</span><span class="p">],</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span>
                               <span class="n">x_length</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span> <span class="n">z_length</span> <span class="o">=</span> <span class="mf">400.0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_columnar</span><span class="p">(</span><span class="s1">&#39;Pyr2&#39;</span><span class="p">,</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">1.0</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">50.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span>
                               <span class="n">min_radius</span> <span class="o">=</span> <span class="mf">50.0</span><span class="p">,</span> <span class="n">max_radius</span> <span class="o">=</span> <span class="mf">200.0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:206311.963
Max density requested:69444.444
Density limit:206311.963
Max density requested:70735.530
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_21_1.png" src="_images/tutorial_cell_placement_21_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">N</span><span class="o">=</span><span class="mi">3000</span>

<span class="n">ctx</span> <span class="o">=</span> <span class="n">CellLocations</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">dmin</span> <span class="o">=</span> <span class="mf">18.0</span>

<span class="n">ctx</span><span class="o">.</span><span class="n">add_positions_ellipsoid</span><span class="p">([</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span><span class="s1">&#39;PV&#39;</span><span class="p">],</span> <span class="n">partitions</span><span class="o">=</span><span class="p">[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">],</span> <span class="n">N</span><span class="o">=</span><span class="n">N</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span> <span class="n">height</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span>
                               <span class="n">x_length</span> <span class="o">=</span> <span class="mf">300.0</span><span class="p">,</span> <span class="n">z_length</span> <span class="o">=</span> <span class="mf">500.0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;prog&#39;</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ctx</span><span class="o">.</span><span class="n">plot_locs</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Density limit:206311.963
Max density requested:127323.954
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_cell_placement_22_1.png" src="_images/tutorial_cell_placement_22_1.png" />
</div>
</div>
<p>Once we are satisfied with the placement, we can pass them to <code class="docutils literal notranslate"><span class="pre">add_nodes()</span></code> as follows (based on the ellipsoid example directly above):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.networks</span> <span class="kn">import</span> <span class="n">NetworkBuilder</span>

<span class="n">net</span> <span class="o">=</span> <span class="n">NetworkBuilder</span><span class="p">(</span><span class="s1">&#39;cortex&#39;</span><span class="p">)</span>
<span class="n">net</span><span class="o">.</span><span class="n">add_nodes</span><span class="p">(</span><span class="n">N</span><span class="o">=</span><span class="n">ctx</span><span class="o">.</span><span class="n">Pyr</span><span class="o">.</span><span class="n">N</span><span class="p">,</span>
              <span class="n">positions</span><span class="o">=</span><span class="n">ctx</span><span class="o">.</span><span class="n">Pyr</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span>
              <span class="n">pop_name</span><span class="o">=</span><span class="s1">&#39;Pyr&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s1">&#39;VisL4&#39;</span><span class="p">,</span> <span class="n">ei</span><span class="o">=</span><span class="s1">&#39;e&#39;</span><span class="p">,</span>  <span class="c1"># optional parameters</span>
              <span class="n">model_type</span><span class="o">=</span><span class="s1">&#39;point_process&#39;</span><span class="p">,</span>  <span class="c1"># Tells the simulator to use point-based neurons</span>
              <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;nest:iaf_psc_alpha&#39;</span><span class="p">,</span>  <span class="c1"># tells the simulator to use NEST iaf_psc_alpha models</span>
              <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;472363762_point.json&#39;</span>  <span class="c1"># File containing iaf_psc_alpha model parameters</span>
             <span class="p">)</span>

<span class="n">net</span><span class="o">.</span><span class="n">add_nodes</span><span class="p">(</span><span class="n">N</span><span class="o">=</span><span class="n">ctx</span><span class="o">.</span><span class="n">PV</span><span class="o">.</span><span class="n">N</span><span class="p">,</span>
              <span class="n">positions</span><span class="o">=</span><span class="n">ctx</span><span class="o">.</span><span class="n">PV</span><span class="o">.</span><span class="n">positions</span><span class="p">,</span>
              <span class="n">pop_name</span><span class="o">=</span><span class="s1">&#39;PV&#39;</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="s1">&#39;VisL4&#39;</span><span class="p">,</span> <span class="n">ei</span><span class="o">=</span><span class="s1">&#39;i&#39;</span><span class="p">,</span>  <span class="c1"># optional parameters</span>
              <span class="n">model_type</span><span class="o">=</span><span class="s1">&#39;point_process&#39;</span><span class="p">,</span>  <span class="c1"># Tells the simulator to use point-based neurons</span>
              <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;nest:iaf_psc_alpha&#39;</span><span class="p">,</span>  <span class="c1"># tells the simulator to use NEST iaf_psc_alpha models</span>
              <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;472912177_point.json&#39;</span>  <span class="c1"># File containing iaf_psc_alpha model parameters</span>
             <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h3><a href="index.html">Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="builder.html">Building Networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulators.html">Simulation Engines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bionet.html">BioNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="pointnet.html">PointNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="popnet.html">PopNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="filternet.html">FilterNet</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analyzer.html">Analyzing the Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Main Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorial_NetworkBuilder_Intro.html">Builder: Using the Network Builder</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_cell_clamped.html">BioNet: Single cell with current injection</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_cell_syn.html">BioNet: Single with with synaptic input</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_pop.html">BioNet: Multiple Nodes with single cell-type</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_multi_pop.html">BioNet: Heterogeneous network</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_pointnet_modeling.html">PointNet: Point-neuron modeling</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_population_modeling.html">PopNet: Population-based firing rate models</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_filter_models.html">FilterNet: Full-field flashing movie</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="ad_tutorials.html">Additional Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Cell placement: Functions for spatially distributing units</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_dynamic_synapses.html">Accessing synapse models in NEST to model dynamic synapses</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_spike_trains_refractory.html">Creating spike trains with refractory periods</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Example Networks</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/AllenInstitute/bmtk.git">Github Profile</a></li>
<li class="toctree-l1"><a class="reference internal" href="workshop.html">Annual Modeling Software Workshop</a></li>
<li class="toctree-l1"><a class="reference internal" href="registration.html">Registration</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributors.html">Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_cite.html">How to Cite</a></li>
<li class="toctree-l1"><a class="reference internal" href="bmtk/bmtk.html">Source Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.analyzer.html">bmtk.analyzer package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.builder.html">bmtk.builder package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.simulator.html">bmtk.simulator package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.tests.html">bmtk.tests package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.utils.html">bmtk.utils package</a></li>
</ul>
</li>
</ul>

<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


    <div class="footer" role="contentinfo">
    </div>
<script type="text/javascript" src="_static/external_assets/javascript/portalFooter.js"></script>

  </body>
</html>