
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Chapter 2: Single cell simulation with external feedfoward input (with BioNet) &#8212; Brain Modeling Toolkit 0.0.6 documentation</title>
    <link rel="stylesheet" href="_static/aibs_sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Chapter 3: Multi-cell, single population network (with BioNet)" href="tutorial_single_pop.html" />
    <link rel="prev" title="Chapter 1: Single cell simulation with current injection (with BioNet)" href="tutorial_single_cell_clamped.html" /> 
  </head><body>
<script type="text/javascript" src="_static/external_assets/javascript/portal.js"></script>
<style>

	/* the icon font-family definition and helper classes */
	@font-face {
	  font-family: 'portal_icon_font';
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677');
	  src: url('_static/external_assets/stylesheets/portal_icon_font.eot?39505677#iefix') format('embedded-opentype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.woff?39505677') format('woff'),
	       url('_static/external_assets/stylesheets/portal_icon_font.ttf?39505677') format('truetype'),
	       url('_static/external_assets/stylesheets/portal_icon_font.svg?39505677#portal_icon_font') format('svg');
	  font-weight: normal;
	  font-style: normal;
	}
	 
	 [class^="icon-"]:before, [class*=" icon-"]:before {
	  font-family: "portal_icon_font";
	  font-style: normal;
	  font-weight: normal;
	  speak: none;
	 
	  display: inline-block;
	  text-decoration: inherit;
	  width: 1em;
	  margin-right: .2em;
	  text-align: center;
	  /* opacity: .8; */
	 
	  /* For safety - reset parent styles, that can break glyph codes*/
	  font-variant: normal;
	  text-transform: none;
	     
	  /* fix buttons height, for twitter bootstrap */
	  line-height: 1em;
	 
	  /* Animation center compensation - margins should be symmetric */
	  /* remove if not needed */
	  margin-left: .2em;
	
	}
	 
	.icon-spin5:before { content: '\e806'; } /* '' */
	.icon-left-open-1:before { content: '\e81e'; } /* '' */
	.icon-plus-1:before { content: '\e801'; } /* '' */
	.icon-facebook:before { content: '\e816'; } /* '' */
	.icon-gplus:before { content: '\e818'; } /* '' */
	.icon-linkedin-3:before { content: '\e815'; } /* '' */
	.icon-twitter:before { content: '\e817'; } /* '' */
	.icon-youtube-play:before { content: '\e805'; } /* '' */
	.icon-plus-2:before { content: '\e809'; } /* '' */
	.icon-record-1:before { content: '\e81c'; } /* '' */
	.icon-cancel:before { content: '\e80b'; } /* '' */
	.icon-search:before { content: '\e804'; } /* '' */
	.icon-cancel-2:before { content: '\e80c'; } /* '' */
	.icon-plus-4:before { content: '\e80d'; } /* '' */
	.icon-cancel-3:before { content: '\e80e'; } /* '' */
	.icon-plus:before { content: '\e808'; } /* '' */
	.icon-down-open:before { content: '\e800'; } /* '' */
	.icon-right-open:before { content: '\e802'; } /* '' */
	.icon-up-open:before { content: '\e803'; } /* '' */
	.icon-stop:before { content: '\e813'; } /* '' */
	.icon-cancel-1:before { content: '\e820'; } /* '' */
	.icon-plus-3:before { content: '\e80a'; } /* '' */
	/* end icon font helpers */

	/* icon font spin animation */
	.animate-spin {
	  -moz-animation: spin 2s infinite linear;
	  -o-animation: spin 2s infinite linear;
	  -webkit-animation: spin 2s infinite linear;
	  animation: spin 2s infinite linear;
	  display: inline-block;
	}
	@-moz-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-webkit-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-o-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@-ms-keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	@keyframes spin {
	  0% {
	    -moz-transform: rotate(0deg);
	    -o-transform: rotate(0deg);
	    -webkit-transform: rotate(0deg);
	    transform: rotate(0deg);
	  }

	  100% {
	    -moz-transform: rotate(359deg);
	    -o-transform: rotate(359deg);
	    -webkit-transform: rotate(359deg);
	    transform: rotate(359deg);
	  }
	}
	/*  end spin animation */	

	@font-face {
	    font-family: 'bebas_neueregular';
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot');
	    src: url('_static/external_assets/stylesheets/bebasneue-webfont.eot?#iefix') format('embedded-opentype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.woff') format('woff'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.ttf') format('truetype'),
	         url('_static/external_assets/stylesheets/bebasneue-webfont.svg#bebas_neueregular') format('svg');
	    font-weight: normal;
	    font-style: normal;
	}

	/* top chrome styles */
	#top_chrome {

		width:100%;
		height:28px;
		background-color:#163E65;
	}

	.chrome_link_block {

		color:#fff !important;	
		display:inline-block;
		height:100%;
		max-height:28px;
	}

	.chrome_link_block:hover {

		background-color:#eee;
		color:#0779BE !important;	
		cursor:pointer;
	}

	#top_chrome a {

		color:inherit !important;	
		text-decoration:none;
	}

	#top_chrome a:hover {

		color:inherit !important;	
	}

	#top_chrome_buttons {

		float:right;
		margin-right:50px;
	}

	.chrome_button {

		font-family:arial,sans-serif;
		font-weight:bold;
		display:inline-block;
		font-size:9px;
		text-align:left;
		text-decoration:none;
		margin:9px 20px 10px 20px;
	}

	.chrome_active {

		background-color:#1b7abd;
		color:#fff;
	}


	/* header element styles*/
	#header_content {

		font-family: bebas_neueregular, arial, sans-serif;
		font-size:22px;
		color:#0779be;		
		letter-spacing:0.03em;

		position:relative;
		width:calc(100% - 112px);
		min-width:800px;
		height:76px;
		border-bottom:2px solid #A5A8AB;
		margin-left:50px;
		margin-right:50px;
        z-index:9999; /* IT-9508 */
	}

	#portal_logo {

		position:absolute;
		top:20px;
		left:0px;
		width:236px;
	}

	#portal_tabs {

		position: absolute;
		bottom: 12px;
		left: 0px;
		margin-bottom: -12px;
		display: block;
	}

	.horizontal_list
	{
		margin: 0;
		padding: 0;
		list-style-type: none;
		display: inline; 		
	}

	.horizontal_list li { 

		display: inline; 	
	}	

	#portal_tab_menu_container {

		display:none;
	}

	#portal_search_input_group {

		position:absolute;
		margin: 0;
		padding: 0;
		bottom:12px;
		right:0px;
		border:1px solid #DAD9D3;
		color:#DAD9D3;
	}

	#portal_search_input_group input {

		margin: 0;
		padding: 0;
		border:none;
		width:110px;
		color:#DAD9D3;
		font-family:Helvetica,sans-serif;
		font-size:12px;
		height:28px;
		margin-left:10px;

		outline:none;
	}

	input::-ms-clear {
		display: none;
	}

	#portal_search_input_group i {

		font-size:0.8em;
		cursor:pointer;
	}

	#pSiteSearchTextInput:focus {

		color:#000;
	}

	.icon_menu_down {

		font-size: 12px;
		margin-left: 4px;
		vertical-align: text-top;
		padding-top: 4px;
		display: inline-block;
	}

	.portal_tab_item {

		background-color: #fff;
		padding: 12px;
		padding-left:20px;
		padding-right:20px;
		display: inline-block !important;		
	}

	.portal_tab_item a:hover {

		color:#fbe300;	
	}

	.portal_tab_item:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}	


	.portal_menu_active {

		background-color: #EEEEEE;
	}	

	.portal_menu_item_active {

		color:#fbe300 !important;	
	}

	.portal_tab_active {

		color:#a5a8ab !important;
	}

	#header_content a {		

		color:inherit !important;
	}

	#header_content a:hover {

		background-color: #0779be;	
		color:#fbe300 !important;	
	}

	#header_content ul {

		list-style: none;
		position: relative;
		top: 58px;
		display: inline-table;
	}

	#header_content ul ul {
		
		display:none;
		opacity:0;
	}

	/* the fly out menus */
	#header_content ul li:hover > ul,
	#header_content ul li.hover > ul {

		position: absolute;
		top: 48px;
		z-index:999;

		display: block;
		width: 152px;
		padding:0px;
		padding-top:20px;
		padding-bottom:10px;
		font-size: 12px;
		font-family: Arial,Sans-Serif;
		font-weight: bold;
		text-transform: uppercase;

		background-color: rgba(8,122,191,0.98);
		color: #fff !important;
		opacity: 1;

		-webkit-animation: fade 0.25s;
		animation: fade 0.25s;
	}

	@-webkit-keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}
	@keyframes fade {
	    from {opacity: 0}
	    to {opacity: 1}
	}	

	#header_content ul li:hover > ul li {

		padding: 10px;
		padding-top: 0px;
		padding-left: 20px;
		display: block;
		opacity:1;
	}

	#header_content ul li:hover > ul a {

		display: block;
		opacity:1;
		white-space:nowrap;
	}

	#portal_menu_start_list {

		left: 81px;

		/* IE sauce */
		left:85px\9;

		/* Safari sauce */
		(;left:79px;);
	}

	#portal_menu_atlases_list {

		left: 290px;
		width: 210px !important;
        height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 303px\9;
        height: 225px\9;

		/* Safari sauce */
		(;left:279px;);
	}

	#portal_menu_tools_list {

		left: 500px;
		width: 172px !important;
		height: 236px;

		padding-bottom:5px !important;

		/* IE sauce */
		left: 513px\9;
		height:225px\9;

		/* Safari sauce */
		(;left:489px;);
		(;height:227px;);
	}

	/* mozilla sauce */
	@-moz-document url-prefix() {

		#header_content {

			height:107px;
		}

		#portal_menu_start_list {

			left:85px;
		}

		#portal_menu_atlases_list {

			left: 302px;		
		}

		#portal_menu_tools_list {

			left: 512px;
			height:236px;
		}
	}


	/* make it look OK on iPad also */
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:portrait) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}  
	}
	@media all and (device-width: 768px) and (device-height: 1024px) and (orientation:landscape) {

		#portal_menu_start_list {

			left:81px;
		}

		#portal_menu_atlases_list {

			left: 291px;		
		}

		#portal_menu_tools_list {

			left: 501px;
			height:236px;
		}	  
	}

	.portal_menu_list_heading {

		color:#9ccae5;
		border-bottom:1px solid #9ccae5;
		padding-bottom:5px !important;
	}

	.portal_menu_list_heading + li {

		padding-top:10px !important;
	}

	.portal_menu_list_heading li {

		padding-bottom:0px;
	}	

	#portal_menu_atlases_list li:last-of-type {

		padding-bottom:15px !important;
	}

	.icon-spin5 {

		color:#000;
	}
</style>
<div id="pHeader">
<div id="top_chrome">
	<div id="top_chrome_buttons">
		<div class="chrome_link_block"><a href="http://www.alleninstitute.org" target="_blank" class="chrome_button">ALLEN INSTITUTE</a></div>
	</div>
</div>

<div id="header_content">

	<a href="http://AllenInstitute.github.io">
		<img id="portal_logo" style="border:0;" src="_static/external_assets/images/AIBS_Logo.png"/>
	</a>
</div>
</div>

<script>

jQuery('#portal_tab_menu').bind('touchstart', function() {

	if(window.console)
		console.log("touchstart");
});
</script>
<link rel="stylesheet" type="text/css" href="_static/external_assets/stylesheets/common_layout.css" />


  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Chapter-2:-Single-cell-simulation-with-external-feedfoward-input-(with-BioNet)">
<h1>Chapter 2: Single cell simulation with external feedfoward input (with BioNet)<a class="headerlink" href="#Chapter-2:-Single-cell-simulation-with-external-feedfoward-input-(with-BioNet)" title="Permalink to this headline">¶</a></h1>
<p>In the previous tutorial we built a single cell and stimulated it with a current injection. In this example we will keep our single-cell network, but instead of stimulation by a step current, we’ll set-up an external network that synapses onto our cell.</p>
<p><strong>Note</strong> - scripts and files for running this tutorial can be found in the directory <a class="reference external" href="sources/chapter02">sources/chapter02/</a></p>
<p><strong>Requirements:</strong> * Python 2.7 or 3.6+ * bmtk * NEURON 7.4+</p>
<div class="section" id="Step-1:-Building-the-network.">
<h2>Step 1: Building the network.<a class="headerlink" href="#Step-1:-Building-the-network." title="Permalink to this headline">¶</a></h2>
<p>Similar to the previous tutorial, we want to build and save a network consisting of a single biophysically detailed cell.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.builder.networks</span> <span class="kn">import</span> <span class="n">NetworkBuilder</span>


<span class="n">cortex</span> <span class="o">=</span> <span class="n">NetworkBuilder</span><span class="p">(</span><span class="s1">&#39;mcortex&#39;</span><span class="p">)</span>
<span class="n">cortex</span><span class="o">.</span><span class="n">add_nodes</span><span class="p">(</span><span class="n">cell_name</span><span class="o">=</span><span class="s1">&#39;Scnn1a_473845048&#39;</span><span class="p">,</span>
                 <span class="n">potental</span><span class="o">=</span><span class="s1">&#39;exc&#39;</span><span class="p">,</span>
                 <span class="n">model_type</span><span class="o">=</span><span class="s1">&#39;biophysical&#39;</span><span class="p">,</span>
                 <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;ctdb:Biophys1.hoc&#39;</span><span class="p">,</span>
                 <span class="n">model_processing</span><span class="o">=</span><span class="s1">&#39;aibs_perisomatic&#39;</span><span class="p">,</span>
                 <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;472363762_fit.json&#39;</span><span class="p">,</span>
                 <span class="n">morphology</span><span class="o">=</span><span class="s1">&#39;Scnn1a_473845048_m.swc&#39;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="n">cortex</span><span class="o">.</span><span class="n">build</span><span class="p">()</span>
<span class="n">cortex</span><span class="o">.</span><span class="n">save_nodes</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch02/network&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>But we will also want a collection of external spike-generating cells that will synapse onto our cell. To do this we create a second network which can represent thalamic input. We will call our network “mthalamus”, and it will consist of 10 cells. These cells are not biophysical but instead “virtual” cells. Virtual cells don’t have a morphology or the normal properties of a neuron, but rather act as spike generators.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="n">thalamus</span> <span class="o">=</span> <span class="n">NetworkBuilder</span><span class="p">(</span><span class="s1">&#39;mthalamus&#39;</span><span class="p">)</span>
<span class="n">thalamus</span><span class="o">.</span><span class="n">add_nodes</span><span class="p">(</span><span class="n">N</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                   <span class="n">pop_name</span><span class="o">=</span><span class="s1">&#39;tON&#39;</span><span class="p">,</span>
                   <span class="n">potential</span><span class="o">=</span><span class="s1">&#39;exc&#39;</span><span class="p">,</span>
                   <span class="n">model_type</span><span class="o">=</span><span class="s1">&#39;virtual&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now that we built our nodes, we want to create a connection between our 10 thalamic cells onto our cortex cell. To do so we use the add_edges function like so:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="n">thalamus</span><span class="o">.</span><span class="n">add_edges</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;pop_name&#39;</span><span class="p">:</span> <span class="s1">&#39;tON&#39;</span><span class="p">},</span> <span class="n">target</span><span class="o">=</span><span class="n">cortex</span><span class="o">.</span><span class="n">nodes</span><span class="p">(),</span>
                   <span class="n">connection_rule</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                   <span class="n">syn_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
                   <span class="n">delay</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
                   <span class="n">weight_function</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                   <span class="n">target_sections</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;basal&#39;</span><span class="p">,</span> <span class="s1">&#39;apical&#39;</span><span class="p">],</span>
                   <span class="n">distance_range</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">],</span>
                   <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;AMPA_ExcToExc.json&#39;</span><span class="p">,</span>
                   <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;exp2syn&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;bmtk.builder.connection_map.ConnectionMap at 0x7f7188ed8910&gt;
</pre></div></div>
</div>
<p>Let us break down how this method call works:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">thalamus</span><span class="o">.</span><span class="n">add_edges</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;pop_name&#39;</span><span class="p">:</span> <span class="s1">&#39;tON&#39;</span><span class="p">},</span> <span class="n">target</span><span class="o">=</span><span class="n">cortex</span><span class="o">.</span><span class="n">nodes</span><span class="p">(),</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Here we specify which set of nodes to use as sources and targets. Our source/pre-synaptic cells are all thamalus cells with the property “pop_name=tON”, which in this case is every thalmus cell (We could also use source=thalamus.nodes(), or source={‘level_of_detail’: ‘filter’}). The target/post-synaptic is all cell(s) of the “cortex” network.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">connection_rule</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
</pre></div>
</div>
<ul class="simple">
<li><p>The connection_rule parameter determines how many synapses exists between every source/target pair. In this very trivial case we are indicating that between every thamalic –&gt; cortical cell connection, there are 5 synapatic connections. In future tutorials we will show how we can create more complex customized rules.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">syn_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
<span class="n">delay</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
<span class="n">weight_function</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Here we are specifying the connection weight. For every connection in this edge-type, there is a connection strength of 5e-05 (units) and a connection delay of 2 ms. The weight function is used to adjust the weights before runtime. Later we will show how to create customized weight functions.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">target_sections</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;basal&#39;</span><span class="p">,</span> <span class="s1">&#39;apical&#39;</span><span class="p">],</span>
<span class="n">distance_range</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">],</span>
</pre></div>
</div>
<ul class="simple">
<li><p>This is used by BioNet to determine where on the post-synaptic cell to place the synapse. By default placement is random within the given section and range.</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;AMPA_ExcToExc.json&#39;</span><span class="p">,</span>
<span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;exp2syn&#39;</span><span class="p">)</span>
</pre></div>
</div>
<ul class="simple">
<li><p>The params_file give the parameters of the synpase, including the time constant and potential. Here we are using an AMPA type synaptic model with an Excitatory connection. The set_params_function is used by BioNet to convert the model into a valid NEURON synaptic object.</p></li>
</ul>
<p>Finally we are ready to build the model and save the thalamic nodes and edges.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="n">thalamus</span><span class="o">.</span><span class="n">build</span><span class="p">()</span>
<span class="n">thalamus</span><span class="o">.</span><span class="n">save_nodes</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch02/network&#39;</span><span class="p">)</span>
<span class="n">thalamus</span><span class="o">.</span><span class="n">save_edges</span><span class="p">(</span><span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch02/network&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The network/ directory will contain multiple nodes and edges files. It should have nodes (and node-types) files for both the thalamus and cortex network. And edges (and edge-types) files for the thalamus –&gt; cortex connections. Nodes and edges for different networks and their connections are spread out across different files which allows us in the future to rebuild, edit or replace part of setup in a piecemeal and efficent manner.</p>
<div class="section" id="Spike-Trains">
<h3>Spike Trains<a class="headerlink" href="#Spike-Trains" title="Permalink to this headline">¶</a></h3>
<p>We need to give our 10 thalamic cells spike trains. There are multiple ways to do this, but an easy way to use a sonata hdf5 file. The following function will create a file to provide the spikes for our 10 cells.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.utils.reports.spike_trains</span> <span class="kn">import</span> <span class="n">PoissonSpikeGenerator</span>

<span class="n">psg</span> <span class="o">=</span> <span class="n">PoissonSpikeGenerator</span><span class="p">(</span><span class="n">population</span><span class="o">=</span><span class="s1">&#39;mthalamus&#39;</span><span class="p">)</span>
<span class="n">psg</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">node_ids</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span>  <span class="c1"># Have 10 nodes to match mthalamus</span>
        <span class="n">firing_rate</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>    <span class="c1"># 10 Hz, we can also pass in a nonhomoegenous function/array</span>
        <span class="n">times</span><span class="o">=</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">3.0</span><span class="p">))</span>    <span class="c1"># Firing starts at 0 s up to 3 s</span>
<span class="n">psg</span><span class="o">.</span><span class="n">to_sonata</span><span class="p">(</span><span class="s1">&#39;sim_ch02/inputs/mthalamus_spikes.h5&#39;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>The spikes files are save as a <a class="reference external" href="https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#spike-file">Sonata formated spikes train</a>. BMTK also allow us to use csv files and NWB files to set the spike train - we just have to tell the configuration file the appropiate format to read.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="n">psg</span><span class="o">.</span><span class="n">to_dataframe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>node_id</th>
      <th>population</th>
      <th>timestamps</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.035761</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.040679</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.096620</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.244019</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.331973</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.386464</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.474339</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.532093</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.641981</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.661213</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.703497</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.773827</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>0.924455</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.018524</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.100684</td>
    </tr>
    <tr>
      <th>15</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.202221</td>
    </tr>
    <tr>
      <th>16</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.236319</td>
    </tr>
    <tr>
      <th>17</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.362414</td>
    </tr>
    <tr>
      <th>18</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.468461</td>
    </tr>
    <tr>
      <th>19</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.518727</td>
    </tr>
    <tr>
      <th>20</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.832532</td>
    </tr>
    <tr>
      <th>21</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.924180</td>
    </tr>
    <tr>
      <th>22</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>1.963368</td>
    </tr>
    <tr>
      <th>23</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.007959</td>
    </tr>
    <tr>
      <th>24</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.041782</td>
    </tr>
    <tr>
      <th>25</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.049386</td>
    </tr>
    <tr>
      <th>26</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.051970</td>
    </tr>
    <tr>
      <th>27</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.064528</td>
    </tr>
    <tr>
      <th>28</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.546912</td>
    </tr>
    <tr>
      <th>29</th>
      <td>0</td>
      <td>mthalamus</td>
      <td>2.590909</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>280</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.196973</td>
    </tr>
    <tr>
      <th>281</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.201703</td>
    </tr>
    <tr>
      <th>282</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.226206</td>
    </tr>
    <tr>
      <th>283</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.264307</td>
    </tr>
    <tr>
      <th>284</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.267023</td>
    </tr>
    <tr>
      <th>285</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.662817</td>
    </tr>
    <tr>
      <th>286</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.696093</td>
    </tr>
    <tr>
      <th>287</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.727064</td>
    </tr>
    <tr>
      <th>288</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.856552</td>
    </tr>
    <tr>
      <th>289</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.869575</td>
    </tr>
    <tr>
      <th>290</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.929920</td>
    </tr>
    <tr>
      <th>291</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>0.985744</td>
    </tr>
    <tr>
      <th>292</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.152974</td>
    </tr>
    <tr>
      <th>293</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.266891</td>
    </tr>
    <tr>
      <th>294</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.390516</td>
    </tr>
    <tr>
      <th>295</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.416569</td>
    </tr>
    <tr>
      <th>296</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.427497</td>
    </tr>
    <tr>
      <th>297</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.563913</td>
    </tr>
    <tr>
      <th>298</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.711437</td>
    </tr>
    <tr>
      <th>299</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.730856</td>
    </tr>
    <tr>
      <th>300</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.796984</td>
    </tr>
    <tr>
      <th>301</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.821040</td>
    </tr>
    <tr>
      <th>302</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.830675</td>
    </tr>
    <tr>
      <th>303</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>1.987896</td>
    </tr>
    <tr>
      <th>304</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.151594</td>
    </tr>
    <tr>
      <th>305</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.313487</td>
    </tr>
    <tr>
      <th>306</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.366667</td>
    </tr>
    <tr>
      <th>307</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.792077</td>
    </tr>
    <tr>
      <th>308</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.825268</td>
    </tr>
    <tr>
      <th>309</th>
      <td>9</td>
      <td>mthalamus</td>
      <td>2.873350</td>
    </tr>
  </tbody>
</table>
<p>310 rows × 3 columns</p>
</div></div>
</div>
</div>
</div>
<div class="section" id="Step-2:-Setting-up-BioNet-environment.">
<h2>Step 2: Setting up BioNet environment.<a class="headerlink" href="#Step-2:-Setting-up-BioNet-environment." title="Permalink to this headline">¶</a></h2>
<div class="section" id="file-structure.">
<h3>file structure.<a class="headerlink" href="#file-structure." title="Permalink to this headline">¶</a></h3>
<p>Before running a simulation, we will need to create the runtime environment, including parameter files, run-script and configuration files. If using the tutorial these files will already be in place, however you should run the following anyway in a command-line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m bmtk.utils.sim_setup  <span class="se">\</span>
   --report-vars v,cai            <span class="se">\</span>
   --network sim_ch02/network     <span class="se">\</span>
   --spikes-inputs mthalamus:sim_ch02/inputs/mthalamus_spikes.h5 <span class="se">\</span>
   --dt <span class="m">0</span>.1             <span class="se">\</span>
   --tstop <span class="m">3000</span>.0       <span class="se">\</span>
   --include-examples   <span class="se">\</span>
   --compile-mechanisms <span class="se">\</span>
   bionet sim_ch02
</pre></div>
</div>
<p>Or call the function directly in python</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.utils.sim_setup</span> <span class="kn">import</span> <span class="n">build_env_bionet</span>

<span class="n">build_env_bionet</span><span class="p">(</span><span class="n">base_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch02&#39;</span><span class="p">,</span>
                 <span class="n">network_dir</span><span class="o">=</span><span class="s1">&#39;sim_ch02/network&#39;</span><span class="p">,</span>
                 <span class="n">tstop</span><span class="o">=</span><span class="mf">3000.0</span><span class="p">,</span> <span class="n">dt</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                 <span class="n">report_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;v&#39;</span><span class="p">,</span> <span class="s1">&#39;cai&#39;</span><span class="p">],</span>     <span class="c1"># Record membrane potential and calcium (default soma)</span>
                 <span class="n">spikes_inputs</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;mthalamus&#39;</span><span class="p">,</span>   <span class="c1"># Name of population which spikes will be generated for</span>
                                <span class="s1">&#39;sim_ch02/inputs/mthalamus_spikes.h5&#39;</span><span class="p">)],</span>
                 <span class="n">include_examples</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>    <span class="c1"># Copies components files</span>
                 <span class="n">compile_mechanisms</span><span class="o">=</span><span class="kc">True</span>   <span class="c1"># Will try to compile NEURON mechanisms</span>
                <span class="p">)</span>
</pre></div>
</div>
</div>
<p>The last thing that we need to do is to update the configuration file to read “thalamus_spikes.csv”. To do so we open simulation_config.json in a text editor and add the following to the <strong>input</strong> section.</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span>&quot;inputs&quot;: {
    &quot;mthalamus_spikes&quot;: {
      &quot;input_type&quot;: &quot;spikes&quot;,
      &quot;module&quot;: &quot;sonata&quot;,
      &quot;input_file&quot;: &quot;${BASE_DIR}/input/mthalamus_spikes.csv&quot;,
      &quot;node_set&quot;: &quot;mthalamus&quot;
    }
}
</pre></div>
</div>
</div>
</div>
<div class="section" id="3.-Running-the-simulation">
<h2>3. Running the simulation<a class="headerlink" href="#3.-Running-the-simulation" title="Permalink to this headline">¶</a></h2>
<p>Once our config file is setup we can run a simulation either through the command line:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python run_bionet.py simulation_config.json
</pre></div>
</div>
<p>or through the script</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.simulator</span> <span class="kn">import</span> <span class="n">bionet</span>


<span class="n">conf</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">Config</span><span class="o">.</span><span class="n">from_json</span><span class="p">(</span><span class="s1">&#39;sim_ch02/simulation_config.json&#39;</span><span class="p">)</span>
<span class="n">conf</span><span class="o">.</span><span class="n">build_env</span><span class="p">()</span>
<span class="n">net</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">BioNetwork</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">)</span>
<span class="n">sim</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">BioSimulator</span><span class="o">.</span><span class="n">from_config</span><span class="p">(</span><span class="n">conf</span><span class="p">,</span> <span class="n">network</span><span class="o">=</span><span class="n">net</span><span class="p">)</span>
<span class="n">sim</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
2019-07-21 13:13:44,864 [INFO] Created log file
Mechanisms already loaded from path: /home/kael/Workspace/bmtk/docs/tutorial/sim_ch02/components/mechanisms.  Aborting.
2019-07-21 13:13:44,899 [INFO] Building cells.
2019-07-21 13:13:45,075 [INFO] Building recurrent connections
2019-07-21 13:13:45,076 [INFO] Building virtual cell stimulations for mthalamus_spikes
2019-07-21 13:13:45,151 [INFO] Running simulation for 3000.000 ms with the time step 0.100 ms
2019-07-21 13:13:45,152 [INFO] Starting timestep: 0 at t_sim: 0.000 ms
2019-07-21 13:13:45,154 [INFO] Block save every 5000 steps
2019-07-21 13:13:45,342 [INFO]     step:5000 t_sim:500.00 ms
2019-07-21 13:13:45,525 [INFO]     step:10000 t_sim:1000.00 ms
2019-07-21 13:13:45,707 [INFO]     step:15000 t_sim:1500.00 ms
2019-07-21 13:13:45,886 [INFO]     step:20000 t_sim:2000.00 ms
2019-07-21 13:13:46,068 [INFO]     step:25000 t_sim:2500.00 ms
2019-07-21 13:13:46,253 [INFO]     step:30000 t_sim:3000.00 ms
2019-07-21 13:13:46,272 [INFO] Simulation completed in 1.121 seconds
</pre></div></div>
</div>
</div>
<div class="section" id="4.-Analyzing-the-run">
<h2>4. Analyzing the run<a class="headerlink" href="#4.-Analyzing-the-run" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.analyzer.spike_trains</span> <span class="kn">import</span> <span class="n">to_dataframe</span>
<span class="n">to_dataframe</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;sim_ch02/simulation_config.json&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>node_ids</th>
      <th>population</th>
      <th>timestamps</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>mcortex</td>
      <td>45.8</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>mcortex</td>
      <td>86.7</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0</td>
      <td>mcortex</td>
      <td>121.9</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0</td>
      <td>mcortex</td>
      <td>205.9</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0</td>
      <td>mcortex</td>
      <td>252.3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0</td>
      <td>mcortex</td>
      <td>287.3</td>
    </tr>
    <tr>
      <th>6</th>
      <td>0</td>
      <td>mcortex</td>
      <td>312.7</td>
    </tr>
    <tr>
      <th>7</th>
      <td>0</td>
      <td>mcortex</td>
      <td>406.3</td>
    </tr>
    <tr>
      <th>8</th>
      <td>0</td>
      <td>mcortex</td>
      <td>483.2</td>
    </tr>
    <tr>
      <th>9</th>
      <td>0</td>
      <td>mcortex</td>
      <td>592.4</td>
    </tr>
    <tr>
      <th>10</th>
      <td>0</td>
      <td>mcortex</td>
      <td>700.0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>0</td>
      <td>mcortex</td>
      <td>734.3</td>
    </tr>
    <tr>
      <th>12</th>
      <td>0</td>
      <td>mcortex</td>
      <td>874.4</td>
    </tr>
    <tr>
      <th>13</th>
      <td>0</td>
      <td>mcortex</td>
      <td>934.9</td>
    </tr>
    <tr>
      <th>14</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1001.2</td>
    </tr>
    <tr>
      <th>15</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1023.7</td>
    </tr>
    <tr>
      <th>16</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1109.1</td>
    </tr>
    <tr>
      <th>17</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1126.3</td>
    </tr>
    <tr>
      <th>18</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1194.6</td>
    </tr>
    <tr>
      <th>19</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1206.0</td>
    </tr>
    <tr>
      <th>20</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1227.3</td>
    </tr>
    <tr>
      <th>21</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1277.4</td>
    </tr>
    <tr>
      <th>22</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1308.4</td>
    </tr>
    <tr>
      <th>23</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1348.6</td>
    </tr>
    <tr>
      <th>24</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1369.2</td>
    </tr>
    <tr>
      <th>25</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1401.5</td>
    </tr>
    <tr>
      <th>26</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1421.0</td>
    </tr>
    <tr>
      <th>27</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1450.0</td>
    </tr>
    <tr>
      <th>28</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1472.8</td>
    </tr>
    <tr>
      <th>29</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1657.1</td>
    </tr>
    <tr>
      <th>30</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1672.0</td>
    </tr>
    <tr>
      <th>31</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1754.7</td>
    </tr>
    <tr>
      <th>32</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1795.6</td>
    </tr>
    <tr>
      <th>33</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1837.3</td>
    </tr>
    <tr>
      <th>34</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1913.4</td>
    </tr>
    <tr>
      <th>35</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1930.0</td>
    </tr>
    <tr>
      <th>36</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1956.9</td>
    </tr>
    <tr>
      <th>37</th>
      <td>0</td>
      <td>mcortex</td>
      <td>1999.6</td>
    </tr>
    <tr>
      <th>38</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2051.5</td>
    </tr>
    <tr>
      <th>39</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2130.1</td>
    </tr>
    <tr>
      <th>40</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2172.8</td>
    </tr>
    <tr>
      <th>41</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2303.4</td>
    </tr>
    <tr>
      <th>42</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2455.7</td>
    </tr>
    <tr>
      <th>43</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2508.4</td>
    </tr>
    <tr>
      <th>44</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2566.1</td>
    </tr>
    <tr>
      <th>45</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2616.4</td>
    </tr>
    <tr>
      <th>46</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2646.7</td>
    </tr>
    <tr>
      <th>47</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2675.6</td>
    </tr>
    <tr>
      <th>48</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2793.9</td>
    </tr>
    <tr>
      <th>49</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2836.1</td>
    </tr>
    <tr>
      <th>50</th>
      <td>0</td>
      <td>mcortex</td>
      <td>2970.5</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">bmtk.analyzer.cell_vars</span> <span class="kn">import</span> <span class="n">plot_report</span>

<span class="n">plot_report</span><span class="p">(</span><span class="n">config_file</span><span class="o">=</span><span class="s1">&#39;sim_ch02/simulation_config.json&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_single_cell_syn_22_0.png" src="_images/tutorial_single_cell_syn_22_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/tutorial_single_cell_syn_22_1.png" src="_images/tutorial_single_cell_syn_22_1.png" />
</div>
</div>
</div>
<div class="section" id="5.-Additional-things-to-do:">
<h2>5. Additional things to do:<a class="headerlink" href="#5.-Additional-things-to-do:" title="Permalink to this headline">¶</a></h2>
<p>When using the Network Builder add_edges method, we gave all the edges the same parameter values (delay, weight, target_section, etc.). All connection created by this method call constitute an single edge-type that share the same parameters, and are specified in the mthalamic_edge_type.csv file</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;network/mthalamus_mcortex_edge_types.csv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>edge_type_id</th>
      <th>target_query</th>
      <th>source_query</th>
      <th>syn_weight</th>
      <th>dynamics_params</th>
      <th>distance_range</th>
      <th>delay</th>
      <th>target_sections</th>
      <th>weight_function</th>
      <th>model_template</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>100</td>
      <td>*</td>
      <td>pop_name=='tON'</td>
      <td>0.001</td>
      <td>AMPA_ExcToExc.json</td>
      <td>[0.0, 150.0]</td>
      <td>2.0</td>
      <td>['basal', 'apical']</td>
      <td>NaN</td>
      <td>exp2syn</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>(if in the build script we called add_edges multiple times, we’d have multiple edge-types).</p>
<p>Using a simple text-editor we can modify this file directly, change parameters before a simulation run without having to rebuild the entire network (although for a network this small it may not be beneficial).</p>
<div class="section" id="weight_function">
<h3>weight_function<a class="headerlink" href="#weight_function" title="Permalink to this headline">¶</a></h3>
<p>By default BioNet uses the value in syn_weight to set a synaptic weight, which is a constant stored in the network files. Often we will want to adjust the synaptic weight between simulations, but don’t want to have to regenerate the network. BioNet allows us to specify custom synaptic weight functions that will calculate synaptic weight before each simulation.</p>
<p>To do so first we must set the value of ‘weight_function’ column. Either we can open up the file mthalamus_mcortex_edge_types.csv with a text-editor and change the column.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 21%" />
<col style="width: 25%" />
<col style="width: 4%" />
<col style="width: 28%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>edge_type_id</p></th>
<th class="head"><p>target_query</p></th>
<th class="head"><p>source_query</p></th>
<th class="head"><p>…</p></th>
<th class="head"><p>weight_function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>100</p></td>
<td><p>*</p></td>
<td><p>pop_name==‘tON’</p></td>
<td><p>…</p></td>
<td><p><em>adjusted_weight</em></p></td>
</tr>
</tbody>
</table>
<p>or we can rebuild the edges</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">thalamus</span><span class="o">.</span><span class="n">add_edges</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;pop_name&#39;</span><span class="p">:</span> <span class="s1">&#39;tON&#39;</span><span class="p">},</span> <span class="n">target</span><span class="o">=</span><span class="n">cortex</span><span class="o">.</span><span class="n">nodes</span><span class="p">(),</span>
                   <span class="n">connection_rule</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                   <span class="n">syn_weight</span><span class="o">=</span><span class="mf">0.001</span><span class="p">,</span>
                   <span class="n">weight_function</span><span class="o">=</span><span class="n">adjusted_weight</span><span class="p">,</span>
                   <span class="n">delay</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
                   <span class="n">target_sections</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;basal&#39;</span><span class="p">,</span> <span class="s1">&#39;apical&#39;</span><span class="p">],</span>
                   <span class="n">distance_range</span><span class="o">=</span><span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">150.0</span><span class="p">],</span>
                   <span class="n">dynamics_params</span><span class="o">=</span><span class="s1">&#39;AMPA_ExcToExc.json&#39;</span><span class="p">,</span>
                   <span class="n">model_template</span><span class="o">=</span><span class="s1">&#39;exp2syn&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then we write a custom weight function. The weight functions will be called during the simulation when building each synapse, and requires three parameters - target_cell, source_cell, and edge_props. These three parameters are dictionaries which can be used to access properties of the source node, target node, and edge, respectively. The function must return a floating point number which will be used to set the synaptic weight</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">adjusted_weights</span><span class="p">(</span><span class="n">target_cell</span><span class="p">,</span> <span class="n">source_cell</span><span class="p">,</span> <span class="n">edge_props</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">target_cell</span><span class="p">[</span><span class="s1">&#39;cell_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Scnn1a&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">edge_prop</span><span class="p">[</span><span class="s2">&quot;weight_max&quot;</span><span class="p">]</span><span class="o">*</span><span class="mf">0.5</span>
    <span class="k">elif</span> <span class="n">target_cell</span><span class="p">[</span><span class="s1">&#39;cell_name&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Rorb&#39;</span>
        <span class="k">return</span> <span class="n">edge_prop</span><span class="p">[</span><span class="s2">&quot;weight_max&quot;</span><span class="p">]</span><span class="o">*</span><span class="mf">1.5</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="o">...</span>
</pre></div>
</div>
<p>Finally we must tell BioNet where to access the function which we can do by using the add_weight_function.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">bmtk.simulator</span> <span class="kn">import</span> <span class="n">bionet</span>

<span class="n">bionet</span><span class="o">.</span><span class="n">nrn</span><span class="o">.</span><span class="n">add_weight_function</span><span class="p">(</span><span class="n">adjusted_weights</span><span class="p">)</span>

<span class="n">conf</span> <span class="o">=</span> <span class="n">bionet</span><span class="o">.</span><span class="n">Config</span><span class="o">.</span><span class="n">from_json</span><span class="p">(</span><span class="s1">&#39;config.json&#39;</span><span class="p">)</span>
<span class="o">...</span>
</pre></div>
</div>
<div class="section" id="Using-NWB-for-spike-trains">
<h4>Using NWB for spike trains<a class="headerlink" href="#Using-NWB-for-spike-trains" title="Permalink to this headline">¶</a></h4>
<p>Instead of using csv files to set the spike trains of our external network, we can also use nwb files. The typical setup would look like the following in the config file:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span>&quot;inputs&quot;: {
    &quot;LGN_spikes&quot;: {
      &quot;input_type&quot;: &quot;spikes&quot;,
      &quot;module&quot;: &quot;nwb&quot;,
      &quot;input_file&quot;: &quot;$INPUT_DIR/lgn_spikes.nwb&quot;,
      &quot;node_set&quot;: &quot;lgn&quot;,
      &quot;trial&quot;: &quot;trial_0&quot;
    },
}
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h3><a href="index.html">Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="builder.html">Building Networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="simulators.html">Simulation Engines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bionet.html">BioNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="pointnet.html">PointNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="popnet.html">PopNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="filternet.html">FilterNet</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="analyzer.html">Analyzing the Results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_NetworkBuilder_Intro.html">Builder: Using the Network Builder</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_cell_clamped.html">BioNet: Single cell with current injection</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">BioNet: Single with with synaptic input</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_single_pop.html">BioNet: Multiple Nodes with single cell-type</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_multi_pop.html">BioNet: Heterogeneous network</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_pointnet_modeling.html">PointNet: Point-neuron modeling</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_population_modeling.html">PopNet: Population-based firing rate models</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_filter_models.html">FilterNet: Full-field flashing movie</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Example Networks</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/AllenInstitute/bmtk.git">Github Profile</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributors.html">Contributors</a></li>
<li class="toctree-l1"><a class="reference internal" href="bmtk/bmtk.html">Source Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.analyzer.html">bmtk.analyzer package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.builder.html">bmtk.builder package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.simulator.html">bmtk.simulator package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.tests.html">bmtk.tests package</a></li>
<li class="toctree-l2"><a class="reference internal" href="bmtk/bmtk.utils.html">bmtk.utils package</a></li>
</ul>
</li>
</ul>

<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


    <div class="footer" role="contentinfo">
    </div>
<script type="text/javascript" src="_static/external_assets/javascript/portalFooter.js"></script>

  </body>
</html>